---
title: "Marm_MicroecoPrep_2025-6-4_SA"
author: "Shayda"
date: "`r Sys.Date()`"
params:

  
---
Chunk errors = knitting error

```{r}
global <- config::get(config = "default")

here::i_am("Marm_MicroecoPrep_2025-6-4_SA.Rmd")
source(here::here(global$setup))
```
# Metadata
```{r}
metadata <- read.table("metadata/metadata.tsv",
                      header = T,
                      sep = "\t")
```

## Sample Subsets
### Both LSK and RAP
```{r}
sample.list.g1 <- metadata %>%
  filter(Group == "Group1") %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.g1 <- sample.list.g1 %>% unlist()

sample.list.g2 <- metadata %>%
  filter(Group == "Group2") %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.g2 <- sample.list.g2 %>% unlist()

sample.list.int <- metadata %>%
  filter(Group_Phase %in% c("gum", "control")) %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.int <- sample.list.int %>% unlist()
```
### LSK
```{r}
sample.list.lsk <- metadata %>%
  filter(R_Primer == "1391R") %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.lsk <- sample.list.lsk %>% unlist()
```
### RAP
```{r}
sample.list.rap <- metadata %>%
  filter(R_Primer == "1492R") %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.rap <- sample.list.rap %>% unlist()

sample.list.rap.g1 <- metadata %>%
  filter(R_Primer == "1492R",
         Group == "Group1") %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.rap.g1 <- sample.list.rap.g1 %>% unlist()

sample.list.rap.g2 <- metadata %>%
  filter(R_Primer == "1492R",
         Group == "Group2") %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.rap.g2 <- sample.list.rap.g2 %>% unlist()

sample.list.rap.int <- metadata %>%
  filter(R_Primer == "1492R",
         Group_Phase %in% c("gum", "control")) %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.rap.int <- sample.list.rap.int %>% unlist()
```
# Alignment
*Generated from the wf-16s workflow*
```{r}
alignment.files.list <- list.files(path = marmoset$read_alignments,
                                   pattern = "*_wf-metagenomics-alignment.csv$",
                                   full.names = TRUE)

alignment.filenames <- list.files(path = marmoset$read_alignments,
                                   pattern = "*_wf-metagenomics-alignment.csv$",
                                   full.names = FALSE)

alignment.files <- lapply(alignment.files.list,
                          read_alignment_file)

alias <- str_remove(alignment.filenames, 
                    "_wf-metagenomics-alignment.csv")
```
```{r}
alignment.files <- Map(function(df, id) {
  if (nrow(df) > 0) {
    df$alias <- id
  } else {
    warning(paste("Data frame for", id, "is empty. Alias not assigned."))
  }
  return(df)
}, alignment.files, alias)
```
*Filter reads based on a minimum coverage of 80*
```{r}
alignments.long <- bind_rows(alignment.files) %>%
                        as_tibble() %>% fix.strings() -> alignments.clean

alignments.long <- dplyr::select(alignments.clean,
                                reference,
                               taxid,
                               species,
                               genus,
                               family,
                               order,
                               class,
                               phylum,
                               superkingdom,
                               Sequence_ID = alias,
                               pc.coverage,
                               n_reads = number.of.reads) %>%
 left_join(dplyr::select(metadata, Sequence_ID), 
            by = join_by(Sequence_ID)) %>%
  dplyr::filter(pc.coverage >= methods_16s$min_cov & !is.na(Sequence_ID)) %>%
  group_by(reference,
           taxid,
           species,
           genus,
           family,
           order,
           class,
           phylum,
           superkingdom,
           Sequence_ID) %>%
  summarize(samp_cov = mean(pc.coverage),
            samp_n_reads = mean(n_reads)) %>% ungroup() %>%
  arrange(Sequence_ID) %>%
  sort.taxa()
```
*Simplify into a table with one row per reference*
```{r}
alignment.refs <- alignments.long %>%
  group_by(reference,
           taxid,
           species,
           genus,
           family,
           order,
           class,
           phylum,
           superkingdom) %>%
  summarize(ref_n_reads = round(sum(samp_n_reads), digits = 2),
            ref_mean_cov = round(mean(samp_cov), digits = 2)) %>% ungroup() %>%
  group_by(species,
           genus,
           family,
           order,
           class,
           phylum,
           superkingdom) %>%
  arrange(species, desc(ref_n_reads), desc(ref_mean_cov)) %>%
                        mutate(ref_order       = row_number(),
                               tax_total_count = sum(ref_n_reads)) %>%
                        ungroup() %>%
                        filter(ref_order == 1) %>%
                        select(-ref_order) %>% sort.taxa()
```
*Format alignment table structure for calculation of relative abundances*
```{r}
alignments    <- alignments.long %>% 
                        group_by(superkingdom,
                                 phylum,
                                 class,
                                 order,
                                 family,
                                 genus,
                                 species,
                                 Sequence_ID) %>%
                          summarize(counts = sum(samp_n_reads)) %>%
                          ungroup() %>%
                          left_join(alignment.refs,
                    by = join_by(superkingdom,
                                 phylum,
                                 class,
                                 order,
                                 family,
                                 genus,
                                 species)) %>%
                    pivot_wider(id_cols    = c(reference,
                                               taxid,
                                               all_of(taxonomy.ordered),
                                               ref_mean_cov),
                                names_from  = Sequence_ID,
                                values_from = counts) %>%
                    sort.taxa() %>%
                    mutate(across(where(is.numeric), ~replace_na(.x, 0)),
                           organism = str_glue("txid{taxid}"))
```
## Sequencing Depth
```{r}
depth <- alignments.long %>%
  filter(!is.na(Sequence_ID)) %>%
  select(Sequence_ID,
         samp_cov,
         samp_n_reads) %>%
  group_by(Sequence_ID) %>%
  summarize(depth         = round(sum(samp_n_reads), digits = 0),
            mean_coverage = round(mean(samp_cov), digits = 2)) %>%
  mutate(mean_coverage = mean_coverage/100) %>%
  ungroup() %>%
  left_join(metadata, by = join_by(Sequence_ID))
```
```{r}
write.csv(depth, "metadata/depth.csv")
```
## Reference Sequences
*Generate an accession reference table for sequences to retrieve*
```{r}
new.refs          <- alignments %>% 
                     select(reference, organism) %>% distinct()

  write.table(new.refs,
              global$tmp_fetch,
              sep       = "\t",
              quote     = FALSE,
              row.names = FALSE,
              col.names = FALSE)
```
*(->Terminal: fetch_refs_cm)*
## FASTA
*Generate fasta to align sequences and assemble phylo tree*
```{r}
new.refseqs <- read.fasta(paste0(global$tmp_fasta3))

write.fasta(sequences = new.refseqs,
            names = names(new.refseqs),
            file.out = global$tmp_fasta4)
```
*(->Terminal: align_tree_cm)*
## Phylogenetic Tree and Representative Fasta
```{r}
tax.tree <- read.tree(marmoset$refseq_tree)
rep.seqs <- read.fasta(marmoset$refseq_fasta)
rep.seqs.tax4fun <- lapply(rep.seqs, function(seq) {gsub("-", "", seq)})
```
# Prepare MicroEco Datasets
## Taxonomy
```{r}
tax.table <- alignments %>%
  column_to_rownames("organism") %>%
  select(all_of(taxonomy.ordered)) %>%
  sort.taxa() %>%
  rename_with(~str_to_title(.x), .cols = everything()) %>%
             mutate(Kingdom = Superkingdom, .keep = "unused") %>%
             relocate(Kingdom) %>%
             distinct() %>%
             tidy_taxonomy()
```
```{r}
taxonomy.list <- alignments %>% 
                 select(organism, taxonomy.ordered) %>% 
                 sort.taxa() %>%
                 rename_with(~str_to_title(.x), .cols = everything()) %>%
                 mutate(Kingdom = Superkingdom, .keep = "unused") %>%
                 relocate(Organism, Kingdom)

write.table(taxonomy.list,
            marmoset$taxonomy_list,
            row.names = FALSE,
            sep = "\t")
```
## OTUs
```{r}
sample.list <- metadata %>%
  distinct(Sequence_ID, Subject, Study_Day) %>%
  arrange(Subject, Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

seqsamples <- sample.list %>% unlist()
```
```{r}
otu.table <- alignments %>% 
          select(organism,
                 any_of(seqsamples)) %>%
                 column_to_rownames("organism")

otu.sample.list <- as.list(names(otu.table))
```
## Sample Table
```{r}
samples.filtered <- setdiff(sample.list, otu.sample.list)

excluded <- metadata %>%
  filter(Sequence_ID %in% samples.filtered)
```
```{r}
sample.table <- metadata %>%
  filter(Sequence_ID %in% otu.sample.list) %>%
  rename_with(~str_replace_all(., " ", "_")) %>%
                rename_with(~str_replace_all(., "-", "_")) %>%
                rename_with(~str_remove_all(., fixed(")"))) %>%
                rename_with(~str_remove_all(., fixed("("))) %>%
                arrange(Study_Day, Subject) %>%
                mutate(day = as.character(Study_Day)) %>%
  column_to_rownames("Sequence_ID") 

write.table(sample.table, marmoset$sample_table)
```
### Merged Sample Table
```{r}
sample.table.merged <- sample.table %>%
  remove_rownames() %>%
  select(all_of(sample_merge_cols)) %>%
  distinct() %>%
  filter(!is.na(Subj_Day)) %>%
  column_to_rownames("Subj_Day") 

write.table(sample.table.merged, marmoset$sample_table)
  
```
## Data Subsets
```{r}
otu.g1 <- otu.table %>%
  select(any_of(samples.g1))
otu.g2 <- otu.table %>%
  select(any_of(samples.g2))
otu.int <- otu.table %>%
  select(any_of(samples.int))
otu.lsk <- otu.table %>%
  select(any_of(samples.lsk))

otu.rap <- otu.table %>%
  select(any_of(samples.rap))
otu.rap.g1 <- otu.table %>%
  select(any_of(samples.rap.g1))
otu.rap.g2 <- otu.table %>%
  select(any_of(samples.rap.g2))
otu.rap.int <- otu.table %>%
  select(any_of(samples.rap.int))

samp.table.g1 <- sample.table %>%
  filter(Group == "Group1")
samp.table.g2 <- sample.table %>%
  filter(Group == "Group2")
samp.table.int <- sample.table %>%
  filter(Group_Phase %in% c("gum", "control"))
samp.table.lsk <- sample.table %>%
  filter(R_Primer == "1391R")

samp.table.rap <- sample.table %>%
  filter(R_Primer == "1492R")
samp.table.rap.g1 <- sample.table %>%
  filter(R_Primer == "1492R",
         Group == "Group1")
samp.table.rap.g2 <- sample.table %>%
  filter(R_Primer == "1492R",
         Group == "Group2")
samp.table.rap.int <- sample.table %>%
  filter(R_Primer == "1492R",
         Group_Phase %in% c("gum", "control"))
```
# Assemble MicroEco Datasets
```{r}
dataset.main <- microtable$new(
  sample_table = sample.table,
  otu_table = otu.table,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.g1 <- microtable$new(
  sample_table = samp.table.g1,
  otu_table = otu.g1,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.g2 <- microtable$new(
  sample_table = samp.table.g2,
  otu_table = otu.g2,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.int <- microtable$new(
  sample_table = samp.table.int,
  otu_table = otu.int,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.lsk <- microtable$new(
  sample_table = samp.table.lsk,
  otu_table = otu.lsk,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.rap <- microtable$new(
  sample_table = samp.table.rap,
  otu_table = otu.rap,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.rap.g1 <- microtable$new(
  sample_table = samp.table.rap.g1,
  otu_table = otu.rap.g1,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.rap.g2 <- microtable$new(
  sample_table = samp.table.rap.g2,
  otu_table = otu.rap.g2,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.rap.int <- microtable$new(
  sample_table = samp.table.rap.int,
  otu_table = otu.rap.int,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)
```
# MicroEco Basic Stats and Data Cleaning
## Rarefaction
*Generate rarefaction plot to determine value*
```{r}
rarefaction_depths <- c(seq.int(0, 10000, by = 500))

rarefaction.main <- trans_rarefy$new(dataset.main,
                                     alphadiv = "Observed",
                                     depth = rarefaction_depths)
rarefaction_plot <- rarefaction.main$plot_rarefy(show_legend = FALSE)
rarefaction_plot
```
```{r}
rarefaction.lsk <- trans_rarefy$new(dataset.lsk,
                                    alphadiv = "Observed",
                                    depth = rarefaction_depths)
rarefaction_lsk <- rarefaction.lsk$plot_rarefy(show_legend = FALSE)
rarefaction_lsk
```
```{r}
rarefaction.rap <- trans_rarefy$new(dataset.rap,
                                    alphadiv = "Observed",
                                    depth = rarefaction_depths)
rarefaction_rap <- rarefaction.rap$plot_rarefy(show_legend = FALSE)
rarefaction_rap
```
```{r}
ggsave("visuals/marm_rarefaction.png", rarefaction_plot)
```
```{r}
ggsave("visuals/marm_lsk_rarefaction.png", rarefaction_lsk)
ggsave("visuals/marm_rap_rarefaction.png", rarefaction_rap)
```
*Rarefy samples (norm = "SRS", sample.size = 4000)*
```{r}
dataset.main$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.g1$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.g2$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.int$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.lsk$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$lsk_rarefy)
dataset.rap$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rap_rarefy)
dataset.rap.g1$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rap_rarefy)
dataset.rap.g2$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rap_rarefy)
dataset.rap.int$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rap_rarefy)
```
## Technical Replicates
*Merge samples for duplicate observations*
```{r}
dataset.main <- dataset.main$merge_samples("Subj_Day")
dataset.g1 <- dataset.g1$merge_samples("Subj_Day")
dataset.g2 <- dataset.g2$merge_samples("Subj_Day")
dataset.int <- dataset.int$merge_samples("Subj_Day")
dataset.lsk <- dataset.lsk$merge_samples("Subj_Day")
dataset.rap <- dataset.rap$merge_samples("Subj_Day")
dataset.rap.g1 <- dataset.rap.g1$merge_samples("Subj_Day")
dataset.rap.g2 <- dataset.rap.g2$merge_samples("Subj_Day")
dataset.rap.int <- dataset.rap.int$merge_samples("Subj_Day")
```
```{r}
dataset.main$sample_table <- data.frame(sample.table.merged[rownames(dataset.main$sample_table), ])
dataset.g1$sample_table <- data.frame(sample.table.merged[rownames(dataset.g1$sample_table), ])
dataset.g2$sample_table <- data.frame(sample.table.merged[rownames(dataset.g2$sample_table), ])
dataset.int$sample_table <- data.frame(sample.table.merged[rownames(dataset.int$sample_table), ])
dataset.lsk$sample_table <- data.frame(sample.table.merged[rownames(dataset.lsk$sample_table), ])
dataset.rap$sample_table <- data.frame(sample.table.merged[rownames(dataset.rap$sample_table), ])
dataset.rap.g1$sample_table <- data.frame(sample.table.merged[rownames(dataset.rap.g1$sample_table), ])
dataset.rap.g2$sample_table <- data.frame(sample.table.merged[rownames(dataset.rap.g2$sample_table), ])
dataset.rap.int$sample_table <- data.frame(sample.table.merged[rownames(dataset.rap.int$sample_table), ])
```
## Relative Abundance
```{r}
dataset.main$cal_abund()
dataset.g1$cal_abund()
dataset.g2$cal_abund()
dataset.int$cal_abund()
dataset.lsk$cal_abund()
dataset.rap$cal_abund()
dataset.rap.g1$cal_abund()
dataset.rap.g2$cal_abund()
dataset.rap.int$cal_abund()
```
### Filter Taxa
*Remove taxa below an abundance of 0.00001 and a frequency of 1*
```{r}
dataset.main$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.g1$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.g2$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.int$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.lsk$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.rap$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.rap.g1$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.rap.g2$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.rap.int$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)
```
## Diversity Metrics
### Alpha Diversity
*Calculate alpha diversity indices (including Faith's Phylogenetic Diversity)*
```{r}
dataset.main$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.g1$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.g2$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.int$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.lsk$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.rap$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.rap.g1$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.rap.g2$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.rap.int$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
```
### Beta Diversity
*Calculate beta diversity (unifrac = TRUE)*
```{r}
dataset.main$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
```{r}
dataset.g1$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
```{r}
dataset.g2$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
```{r}
dataset.int$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
```{r}
dataset.lsk$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
```{r}
dataset.rap$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
```{r}
dataset.rap.g1$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
```{r}
dataset.rap.g2$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
```{r}
dataset.rap.int$cal_betadiv(unifrac = methods_16s$marm$unifrac)
```
## Export Main Datasets
```{r}
dataset.main$save_table(marmoset$microeco$data$main, sep = "\t")
dataset.g1$save_table(marmoset$microeco$data$g1, sep = "\t")
dataset.g2$save_table(marmoset$microeco$data$g2, sep = "\t")
dataset.int$save_table(marmoset$microeco$data$int, sep = "\t")
dataset.lsk$save_table(marmoset$microeco$data$lsk, sep = "\t")
dataset.rap$save_table(marmoset$microeco$data$rap, sep = "\t")
dataset.rap.g1$save_table(marmoset$microeco$data$rap_g1, sep = "\t")
dataset.rap.g2$save_table(marmoset$microeco$data$rap_g2, sep = "\t")
dataset.rap.int$save_table(marmoset$microeco$data$rap_int, sep = "\t")

dataset.main$save_abund(marmoset$microeco$abund$main, sep = "\t")
dataset.g1$save_abund(marmoset$microeco$abund$g1, sep = "\t")
dataset.g2$save_abund(marmoset$microeco$abund$g2, sep = "\t")
dataset.int$save_abund(marmoset$microeco$abund$int, sep = "\t")
dataset.lsk$save_abund(marmoset$microeco$abund$lsk, sep = "\t")
dataset.rap$save_abund(marmoset$microeco$abund$rap, sep = "\t")
dataset.rap.g1$save_abund(marmoset$microeco$abund$rap_g1, sep = "\t")
dataset.rap.g2$save_abund(marmoset$microeco$abund$rap_g2, sep = "\t")
dataset.rap.int$save_abund(marmoset$microeco$abund$rap_int, sep = "\t")

dataset.main$save_alphadiv(marmoset$microeco$alpha$main)
dataset.g1$save_alphadiv(marmoset$microeco$alpha$g1)
dataset.g2$save_alphadiv(marmoset$microeco$alpha$g2)
dataset.int$save_alphadiv(marmoset$microeco$alpha$int)
dataset.lsk$save_alphadiv(marmoset$microeco$alpha$lsk)
dataset.rap$save_alphadiv(marmoset$microeco$alpha$rap)
dataset.rap.g1$save_alphadiv(marmoset$microeco$alpha$rap_g1)
dataset.rap.g2$save_alphadiv(marmoset$microeco$alpha$rap_g2)
dataset.rap.int$save_alphadiv(marmoset$microeco$alpha$rap_int)

dataset.main$save_betadiv(marmoset$microeco$beta$main)
dataset.g1$save_betadiv(marmoset$microeco$beta$g1)
dataset.g2$save_betadiv(marmoset$microeco$beta$g2)
dataset.int$save_betadiv(marmoset$microeco$beta$int)
dataset.lsk$save_betadiv(marmoset$microeco$beta$lsk)
dataset.rap$save_betadiv(marmoset$microeco$beta$rap)
dataset.rap.g1$save_betadiv(marmoset$microeco$beta$rap_g1)
dataset.rap.g2$save_betadiv(marmoset$microeco$beta$rap_g2)
dataset.rap.int$save_betadiv(marmoset$microeco$beta$rap_int)
```
## Functional Profiles
### KEGG
```{r}
dataset.keg.main <- microtable$new(
                               sample_table = sample.table,
                               otu_table    = otu.table,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.g1 <- microtable$new(
                               sample_table = samp.table.g1,
                               otu_table    = otu.g1,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.g2 <- microtable$new(
                               sample_table = samp.table.g2,
                               otu_table    = otu.g2,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.int <- microtable$new(
                               sample_table = samp.table.int,
                               otu_table    = otu.int,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)

dataset.keg.lsk <- microtable$new(
                               sample_table = samp.table.lsk,
                               otu_table    = otu.lsk,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.rap <- microtable$new(
                               sample_table = samp.table.rap,
                               otu_table    = otu.rap,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.rap.g1 <- microtable$new(
                               sample_table = samp.table.rap.g1,
                               otu_table    = otu.rap.g1,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.rap.g2 <- microtable$new(
                               sample_table = samp.table.rap.g2,
                               otu_table    = otu.rap.g2,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.rap.int <- microtable$new(
                               sample_table = samp.table.rap.int,
                               otu_table    = otu.rap.int,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
```
```{r}
dataset.keg.main$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.keg.g1$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.keg.g2$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.keg.int$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.keg.lsk$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$lsk_rarefy)
dataset.keg.rap$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rap_rarefy)
dataset.keg.rap.g1$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rap_rarefy)
dataset.keg.rap.g2$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rap_rarefy)
dataset.keg.rap.int$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rap_rarefy)
```
```{r}
dataset.keg.main <- dataset.keg.main$merge_samples("Subj_Day")
dataset.keg.g1 <- dataset.keg.g1$merge_samples("Subj_Day")
dataset.keg.g2 <- dataset.keg.g2$merge_samples("Subj_Day")
dataset.keg.int <- dataset.keg.int$merge_samples("Subj_Day")
dataset.keg.lsk <- dataset.keg.lsk$merge_samples("Subj_Day")
dataset.keg.rap <- dataset.keg.rap$merge_samples("Subj_Day")
dataset.keg.rap.g1 <- dataset.keg.rap.g1$merge_samples("Subj_Day")
dataset.keg.rap.g2 <- dataset.keg.rap.g2$merge_samples("Subj_Day")
dataset.keg.rap.int <- dataset.keg.rap.int$merge_samples("Subj_Day")
```
```{r}
dataset.keg.main$sample_table <- data.frame(dataset.keg.main$sample_table, sample.table.merged[rownames(dataset.keg.main$sample_table), ])
dataset.keg.g1$sample_table <- data.frame(dataset.keg.g1$sample_table, sample.table.merged[rownames(dataset.keg.g1$sample_table), ])
dataset.keg.g2$sample_table <- data.frame(dataset.keg.g2$sample_table, sample.table.merged[rownames(dataset.keg.g2$sample_table), ])
dataset.keg.int$sample_table <- data.frame(dataset.keg.int$sample_table, sample.table.merged[rownames(dataset.keg.int$sample_table), ])
dataset.keg.lsk$sample_table <- data.frame(dataset.keg.lsk$sample_table, sample.table.merged[rownames(dataset.keg.lsk$sample_table), ])
dataset.keg.rap$sample_table <- data.frame(dataset.keg.rap$sample_table, sample.table.merged[rownames(dataset.keg.rap$sample_table), ])
dataset.keg.rap.g1$sample_table <- data.frame(dataset.keg.rap.g1$sample_table, sample.table.merged[rownames(dataset.keg.rap.g1$sample_table), ])
dataset.keg.rap.g2$sample_table <- data.frame(dataset.keg.rap.g2$sample_table, sample.table.merged[rownames(dataset.keg.rap.g2$sample_table), ])
dataset.keg.rap.int$sample_table <- data.frame(dataset.keg.rap.int$sample_table, sample.table.merged[rownames(dataset.keg.rap.int$sample_table), ])
```
```{r}
dataset.keg.main$cal_abund()
dataset.keg.g1$cal_abund()
dataset.keg.g2$cal_abund()
dataset.keg.int$cal_abund()
dataset.keg.lsk$cal_abund()
dataset.keg.rap$cal_abund()
dataset.keg.rap.g1$cal_abund()
dataset.keg.rap.g2$cal_abund()
dataset.keg.rap.int$cal_abund()
```
```{r}
dataset.keg.main$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.g1$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.g2$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.int$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.lsk$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.rap$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.rap.g1$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.rap.g2$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.rap.int$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
```
```{r}
keg.main <- trans_func$new(dataset.keg.main)
keg.g1 <- trans_func$new(dataset.keg.g1)
keg.g2 <- trans_func$new(dataset.keg.g2)
keg.int <- trans_func$new(dataset.keg.int)
keg.lsk <- trans_func$new(dataset.keg.lsk)
keg.rap <- trans_func$new(dataset.keg.rap)
keg.rap.g1 <- trans_func$new(dataset.keg.rap.g1)
keg.rap.g2 <- trans_func$new(dataset.keg.rap.g2)
keg.rap.int <- trans_func$new(dataset.keg.rap.int)

```
```{r}
keg.main$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.g1$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.g2$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.int$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)

keg.lsk$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.rap$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.rap.g1$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.rap.g2$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.rap.int$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
```
```{r}
data("Tax4Fun2_KEGG")
```
#### Main
```{r}
keg.main$cal_tax4fun2_FRI()
```
```{r}
keg.data.main <- microtable$new(
  otu_table = keg.main$res_tax4fun2_pathway,
  tax_table = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.main$sample_table)

keg.data.main$tidy_dataset()

dataset.keg.main$sample_table <- data.frame(dataset.keg.main$sample_table, sample.table.merged[rownames(dataset.keg.main$sample_table), ])

keg.data.main$cal_abund()
keg.data.main$cal_alphadiv()
keg.data.main$cal_betadiv()
keg.data.main$save_table(marmoset$microeco$data$keg$main, sep = "\t")
keg.data.main$save_abund(marmoset$microeco$abund$keg$main, sep = "\t")
keg.data.main$save_alphadiv(marmoset$microeco$alpha$keg$main)
keg.data.main$save_betadiv(marmoset$microeco$beta$keg$main)
```
#### Group 1
```{r}
keg.g1$cal_tax4fun2_FRI()
```
```{r}
keg.data.g1 <- microtable$new(
  otu_table = keg.g1$res_tax4fun2_pathway,
  tax_table = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.g1$sample_table)

keg.data.g1$tidy_dataset()

dataset.keg.g1$sample_table <- data.frame(dataset.keg.g1$sample_table, sample.table.merged[rownames(dataset.keg.g1$sample_table), ])

keg.data.g1$cal_abund()
keg.data.g1$cal_alphadiv()
keg.data.g1$cal_betadiv()
keg.data.g1$save_table(marmoset$microeco$data$keg$g1, sep = "\t")
keg.data.g1$save_abund(marmoset$microeco$abund$keg$g1, sep = "\t")
keg.data.g1$save_alphadiv(marmoset$microeco$alpha$keg$g1)
keg.data.g1$save_betadiv(marmoset$microeco$beta$keg$g1)
```
#### Group 2
```{r}
keg.g2$cal_tax4fun2_FRI()
```
```{r}
keg.data.g2 <- microtable$new(
  otu_table    = keg.g2$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.g2$sample_table)

keg.data.g2$tidy_dataset()

dataset.keg.g2$sample_table <- data.frame(dataset.keg.g2$sample_table, sample.table.merged[rownames(dataset.keg.g2$sample_table), ])

keg.data.g2$cal_abund()
keg.data.g2$cal_alphadiv()
keg.data.g2$cal_betadiv()
keg.data.g2$save_table(marmoset$microeco$data$keg$g2, sep = "\t")
keg.data.g2$save_abund(marmoset$microeco$abund$keg$g2, sep = "\t")
keg.data.g2$save_alphadiv(marmoset$microeco$alpha$keg$g2)
keg.data.g2$save_betadiv(marmoset$microeco$beta$keg$g2)
```
#### Intervention
```{r}
keg.int$cal_tax4fun2_FRI()
```
```{r}
keg.data.int <- microtable$new(
  otu_table    = keg.int$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.int$sample_table)

keg.data.int$tidy_dataset()

dataset.keg.int$sample_table <- data.frame(dataset.keg.int$sample_table, sample.table.merged[rownames(dataset.keg.int$sample_table), ])

keg.data.int$cal_abund()
keg.data.int$cal_alphadiv()
keg.data.int$cal_betadiv()
keg.data.int$save_table(marmoset$microeco$data$keg$int, sep = "\t")
keg.data.int$save_abund(marmoset$microeco$abund$keg$int, sep = "\t")
keg.data.int$save_alphadiv(marmoset$microeco$alpha$keg$int)
keg.data.int$save_betadiv(marmoset$microeco$beta$keg$int)
```
#### LSK
```{r}
keg.lsk$cal_tax4fun2_FRI()
```
```{r}
keg.data.lsk <- microtable$new(
  otu_table    = keg.lsk$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.lsk$sample_table)

keg.data.lsk$tidy_dataset()

dataset.keg.lsk$sample_table <- data.frame(dataset.keg.lsk$sample_table, sample.table.merged[rownames(dataset.keg.lsk$sample_table), ])

keg.data.lsk$cal_abund()
keg.data.lsk$cal_alphadiv()
keg.data.lsk$cal_betadiv()
keg.data.lsk$save_table(marmoset$microeco$data$keg$lsk, sep = "\t")
keg.data.lsk$save_abund(marmoset$microeco$abund$keg$lsk, sep = "\t")
keg.data.lsk$save_alphadiv(marmoset$microeco$alpha$keg$lsk)
keg.data.lsk$save_betadiv(marmoset$microeco$beta$keg$lsk)
```
#### RAP
```{r}
keg.rap$cal_tax4fun2_FRI()
```
```{r}
keg.data.rap <- microtable$new(
  otu_table    = keg.rap$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.rap$sample_table)

keg.data.rap$tidy_dataset()

dataset.keg.rap$sample_table <- data.frame(dataset.keg.rap$sample_table, sample.table.merged[rownames(dataset.keg.rap$sample_table), ])

keg.data.rap$cal_abund()
keg.data.rap$cal_alphadiv()
keg.data.rap$cal_betadiv()
keg.data.rap$save_table(marmoset$microeco$data$keg$rap, sep = "\t")
keg.data.rap$save_abund(marmoset$microeco$abund$keg$rap, sep = "\t")
keg.data.rap$save_alphadiv(marmoset$microeco$alpha$keg$rap)
keg.data.rap$save_betadiv(marmoset$microeco$beta$keg$rap)
```
```{r}
keg.rap.g1$cal_tax4fun2_FRI()
```
```{r}
keg.data.rap.g1 <- microtable$new(
  otu_table    = keg.rap.g1$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.rap.g1$sample_table)

keg.data.rap.g1$tidy_dataset()

dataset.keg.rap.g1$sample_table <- data.frame(dataset.keg.rap.g1$sample_table, sample.table.merged[rownames(dataset.keg.rap.g1$sample_table), ])

keg.data.rap.g1$cal_abund()
keg.data.rap.g1$cal_alphadiv()
keg.data.rap.g1$cal_betadiv()
keg.data.rap.g1$save_table(marmoset$microeco$data$keg$rap_g1, sep = "\t")
keg.data.rap.g1$save_abund(marmoset$microeco$abund$keg$rap_g1, sep = "\t")
keg.data.rap.g1$save_alphadiv(marmoset$microeco$alpha$keg$rap_g1)
keg.data.rap.g1$save_betadiv(marmoset$microeco$beta$keg$rap_g1)
```
```{r}
keg.rap.g2$cal_tax4fun2_FRI()
```
```{r}
keg.data.rap.g2 <- microtable$new(
  otu_table    = keg.rap.g2$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.rap.g2$sample_table)

keg.data.rap.g2$tidy_dataset()

dataset.keg.rap.g2$sample_table <- data.frame(dataset.keg.rap.g2$sample_table, sample.table.merged[rownames(dataset.keg.rap.g2$sample_table), ])

keg.data.rap.g2$cal_abund()
keg.data.rap.g2$cal_alphadiv()
keg.data.rap.g2$cal_betadiv()
keg.data.rap.g2$save_table(marmoset$microeco$data$keg$rap_g2, sep = "\t")
keg.data.rap.g2$save_abund(marmoset$microeco$abund$keg$rap_g2, sep = "\t")
keg.data.rap.g2$save_alphadiv(marmoset$microeco$alpha$keg$rap_g2)
keg.data.rap.g2$save_betadiv(marmoset$microeco$beta$keg$rap_g2)
```
```{r}
keg.rap.int$cal_tax4fun2_FRI()
```
```{r}
keg.data.rap.int <- microtable$new(
  otu_table    = keg.rap.int$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.rap.int$sample_table)

keg.data.rap.int$tidy_dataset()

dataset.keg.rap.int$sample_table <- data.frame(dataset.keg.rap.int$sample_table, sample.table.merged[rownames(dataset.keg.rap.int$sample_table), ])

keg.data.rap.int$cal_abund()
keg.data.rap.int$cal_alphadiv()
keg.data.rap.int$cal_betadiv()
keg.data.rap.int$save_table(marmoset$microeco$data$keg$rap_int, sep = "\t")
keg.data.rap.int$save_abund(marmoset$microeco$abund$keg$rap_int, sep = "\t")
keg.data.rap.int$save_alphadiv(marmoset$microeco$alpha$keg$rap_int)
keg.data.rap.int$save_betadiv(marmoset$microeco$beta$keg$rap_int)
```
### FAPROTAX
```{r}
fpt.main        <- trans_func$new(dataset.main)
fpt.g1          <- trans_func$new(dataset.g1)
fpt.g2          <- trans_func$new(dataset.g2)
fpt.int         <- trans_func$new(dataset.int)
fpt.lsk <- trans_func$new(dataset.lsk)
fpt.rap <- trans_func$new(dataset.rap)
fpt.rap.g1 <- trans_func$new(dataset.rap.g1)
fpt.rap.g2 <- trans_func$new(dataset.rap.g2)
fpt.rap.int <- trans_func$new(dataset.rap.int)

fpt.main$cal_spe_func(prok_database = "FAPROTAX")
fpt.g1$cal_spe_func(prok_database = "FAPROTAX")
fpt.g2$cal_spe_func(prok_database = "FAPROTAX")
fpt.int$cal_spe_func(prok_database = "FAPROTAX")
fpt.lsk$cal_spe_func(prok_database = "FAPROTAX")
fpt.rap$cal_spe_func(prok_database = "FAPROTAX")
fpt.rap.g1$cal_spe_func(prok_database = "FAPROTAX")
fpt.rap.g2$cal_spe_func(prok_database = "FAPROTAX")
fpt.rap.int$cal_spe_func(prok_database = "FAPROTAX")

fpt.main$cal_spe_func_perc(abundance_weighted = T)
fpt.g1$cal_spe_func_perc(abundance_weighted = T)
fpt.g2$cal_spe_func_perc(abundance_weighted = T)
fpt.int$cal_spe_func_perc(abundance_weighted = T)
fpt.lsk$cal_spe_func_perc(abundance_weighted = T)
fpt.rap$cal_spe_func_perc(abundance_weighted = T)
fpt.rap.g1$cal_spe_func_perc(abundance_weighted = T)
fpt.rap.g2$cal_spe_func_perc(abundance_weighted = T)
fpt.rap.int$cal_spe_func_perc(abundance_weighted = T)
```
```{r}
fpt.main$trans_spe_func_perc()
fpt.g1$trans_spe_func_perc()
fpt.g2$trans_spe_func_perc()
fpt.int$trans_spe_func_perc()
fpt.lsk$trans_spe_func_perc()
fpt.rap$trans_spe_func_perc()
fpt.rap.g1$trans_spe_func_perc()
fpt.rap.g2$trans_spe_func_perc()
fpt.rap.int$trans_spe_func_perc()

fpt.main.df        <- as.data.frame(t(fpt.main$res_spe_func_perc), check.names = F)
fpt.g1.df          <- as.data.frame(t(fpt.g1$res_spe_func_perc), check.names = F)
fpt.g2.df          <- as.data.frame(t(fpt.g2$res_spe_func_perc), check.names = F)
fpt.int.df         <- as.data.frame(t(fpt.int$res_spe_func_perc), check.names = F)
fpt.lsk.df <- as.data.frame(t(fpt.lsk$res_spe_func_perc), check.names = F)
fpt.rap.df <- as.data.frame(t(fpt.rap$res_spe_func_perc), check.names = F)
fpt.rap.g1.df <- as.data.frame(t(fpt.rap.g1$res_spe_func_perc), check.names = F)
fpt.rap.g2.df <- as.data.frame(t(fpt.rap.g2$res_spe_func_perc), check.names = F)
fpt.rap.int.df <- as.data.frame(t(fpt.rap.int$res_spe_func_perc), check.names = F)

fpt.main.tax        <- as.data.frame(rownames(fpt.main.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.g1.tax          <- as.data.frame(rownames(fpt.g1.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.g2.tax          <- as.data.frame(rownames(fpt.g2.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.int.tax         <- as.data.frame(rownames(fpt.int.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.lsk.tax  <- as.data.frame(rownames(fpt.lsk.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.rap.tax  <- as.data.frame(rownames(fpt.rap.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.rap.g1.tax  <- as.data.frame(rownames(fpt.rap.g1.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.rap.g2.tax  <- as.data.frame(rownames(fpt.rap.g2.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.rap.int.tax  <- as.data.frame(rownames(fpt.rap.int.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))

fpt.main.otu        <- fpt.main.df %>% rownames_to_column("Function") %>% left_join(fpt.main.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.g1.otu          <- fpt.g1.df %>% rownames_to_column("Function") %>% left_join(fpt.g1.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.g2.otu          <- fpt.g2.df %>% rownames_to_column("Function") %>% left_join(fpt.g2.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.int.otu         <- fpt.int.df %>% rownames_to_column("Function") %>% left_join(fpt.int.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.lsk.otu <- fpt.lsk.df %>% rownames_to_column("Function") %>% left_join(fpt.lsk.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.rap.otu <- fpt.rap.df %>% rownames_to_column("Function") %>% left_join(fpt.rap.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.rap.g1.otu <- fpt.rap.g1.df %>% rownames_to_column("Function") %>% left_join(fpt.rap.g1.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.rap.g2.otu <- fpt.rap.g2.df %>% rownames_to_column("Function") %>% left_join(fpt.rap.g2.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.rap.int.otu <- fpt.rap.int.df %>% rownames_to_column("Function") %>% left_join(fpt.rap.int.tax) %>% column_to_rownames("func_name") %>% select(-Function)

fpt.main.tax        <- fpt.main.tax %>% column_to_rownames("func_name")
fpt.g1.tax          <- fpt.g1.tax %>% column_to_rownames("func_name")
fpt.g2.tax          <- fpt.g2.tax %>% column_to_rownames("func_name")
fpt.int.tax         <- fpt.int.tax %>% column_to_rownames("func_name")
fpt.lsk.tax <- fpt.lsk.tax %>% column_to_rownames("func_name")
fpt.rap.tax <- fpt.rap.tax %>% column_to_rownames("func_name")
fpt.rap.g1.tax <- fpt.rap.g1.tax %>% column_to_rownames("func_name")
fpt.rap.g2.tax <- fpt.rap.g2.tax %>% column_to_rownames("func_name")
fpt.rap.int.tax <- fpt.rap.int.tax %>% column_to_rownames("func_name")
```
```{r}
fpt.data.main <- microtable$new(otu_table    = fpt.main.otu, 
                                tax_table    = fpt.main.tax, 
                                sample_table = dataset.main$sample_table)

fpt.data.g1   <- microtable$new(otu_table    = fpt.g1.otu, 
                                tax_table    = fpt.g1.tax, 
                                sample_table = dataset.g1$sample_table)

fpt.data.g2   <- microtable$new(otu_table    = fpt.g2.otu, 
                                tax_table    = fpt.g2.tax, 
                                sample_table = dataset.g2$sample_table)

fpt.data.int  <- microtable$new(otu_table    = fpt.int.otu, 
                                tax_table    = fpt.int.tax, 
                                sample_table = dataset.int$sample_table)

fpt.data.lsk  <- microtable$new(otu_table    = fpt.lsk.otu, 
                                tax_table    = fpt.lsk.tax, 
                                sample_table = dataset.lsk$sample_table)

fpt.data.rap  <- microtable$new(otu_table    = fpt.rap.otu, 
                                tax_table    = fpt.rap.tax, 
                                sample_table = dataset.rap$sample_table)

fpt.data.rap.g1  <- microtable$new(otu_table    = fpt.rap.g1.otu, 
                                tax_table    = fpt.rap.g1.tax, 
                                sample_table = dataset.rap.g1$sample_table)

fpt.data.rap.g2  <- microtable$new(otu_table    = fpt.rap.g2.otu, 
                                tax_table    = fpt.rap.g2.tax, 
                                sample_table = dataset.rap.g2$sample_table)

fpt.data.rap.int  <- microtable$new(otu_table    = fpt.rap.int.otu, 
                                tax_table    = fpt.rap.int.tax, 
                                sample_table = dataset.rap.int$sample_table)

fpt.data.main$tidy_dataset()
fpt.data.g1$tidy_dataset()
fpt.data.g2$tidy_dataset()
fpt.data.int$tidy_dataset()
fpt.data.lsk$tidy_dataset()
fpt.data.rap$tidy_dataset()
fpt.data.rap.g1$tidy_dataset()
fpt.data.rap.g2$tidy_dataset()
fpt.data.rap.int$tidy_dataset()

fpt.data.main$cal_abund()
fpt.data.g1$cal_abund()
fpt.data.g2$cal_abund()
fpt.data.int$cal_abund()
fpt.data.lsk$cal_abund()
fpt.data.rap$cal_abund()
fpt.data.rap.g1$cal_abund()
fpt.data.rap.g2$cal_abund()
fpt.data.rap.int$cal_abund()

fpt.data.main$cal_alphadiv()
fpt.data.g1$cal_alphadiv()
fpt.data.g2$cal_alphadiv()
fpt.data.int$cal_alphadiv()
fpt.data.lsk$cal_alphadiv()
fpt.data.rap$cal_alphadiv()
fpt.data.rap.g1$cal_alphadiv()
fpt.data.rap.g2$cal_alphadiv()
fpt.data.rap.int$cal_alphadiv()

fpt.data.main$cal_betadiv()
fpt.data.g1$cal_betadiv()
fpt.data.g2$cal_betadiv()
fpt.data.int$cal_betadiv()
fpt.data.lsk$cal_betadiv()
fpt.data.rap$cal_betadiv()
fpt.data.rap.g1$cal_betadiv()
fpt.data.rap.g2$cal_betadiv()
fpt.data.rap.int$cal_betadiv()
```
```{r}
fpt.data.main$save_table(marmoset$microeco$data$fpt$main, sep = "\t")
fpt.data.g1$save_table(marmoset$microeco$data$fpt$g1, sep = "\t")
fpt.data.g2$save_table(marmoset$microeco$data$fpt$g2, sep = "\t")
fpt.data.int$save_table(marmoset$microeco$data$fpt$int, sep = "\t")
fpt.data.lsk$save_table(marmoset$microeco$data$fpt$lsk, sep = "\t")
fpt.data.rap$save_table(marmoset$microeco$data$fpt$rap, sep = "\t")
fpt.data.rap.g1$save_table(marmoset$microeco$data$fpt$rap_g1, sep = "\t")
fpt.data.rap.g2$save_table(marmoset$microeco$data$fpt$rap_g2, sep = "\t")
fpt.data.rap.int$save_table(marmoset$microeco$data$fpt$rap_int, sep = "\t")

fpt.data.main$save_abund(marmoset$microeco$abund$fpt$main, sep = "\t")
fpt.data.g1$save_abund(marmoset$microeco$abund$fpt$g1, sep = "\t")
fpt.data.g2$save_abund(marmoset$microeco$abund$fpt$g2, sep = "\t")
fpt.data.int$save_abund(marmoset$microeco$abund$fpt$int, sep = "\t")
fpt.data.lsk$save_abund(marmoset$microeco$abund$fpt$lsk, sep = "\t")
fpt.data.rap$save_abund(marmoset$microeco$abund$fpt$rap, sep = "\t")
fpt.data.rap.g1$save_abund(marmoset$microeco$abund$fpt$rap_g1, sep = "\t")
fpt.data.rap.g2$save_abund(marmoset$microeco$abund$fpt$rap_g2, sep = "\t")
fpt.data.rap.int$save_abund(marmoset$microeco$abund$fpt$rap_int, sep = "\t")

fpt.data.main$save_alphadiv(marmoset$microeco$alpha$fpt$main)
fpt.data.g1$save_alphadiv(marmoset$microeco$alpha$fpt$g1)
fpt.data.g2$save_alphadiv(marmoset$microeco$alpha$fpt$g2)
fpt.data.int$save_alphadiv(marmoset$microeco$alpha$fpt$int)
fpt.data.lsk$save_alphadiv(marmoset$microeco$alpha$fpt$lsk)
fpt.data.rap$save_alphadiv(marmoset$microeco$alpha$fpt$rap)
fpt.data.rap.g1$save_alphadiv(marmoset$microeco$alpha$fpt$rap_g1)
fpt.data.rap.g2$save_alphadiv(marmoset$microeco$alpha$fpt$rap_g2)
fpt.data.rap.int$save_alphadiv(marmoset$microeco$alpha$fpt$rap_int)

fpt.data.main$save_betadiv(marmoset$microeco$beta$fpt$main)
fpt.data.g1$save_betadiv(marmoset$microeco$beta$fpt$g1)
fpt.data.g2$save_betadiv(marmoset$microeco$beta$fpt$g2)
fpt.data.int$save_betadiv(marmoset$microeco$beta$fpt$int)
fpt.data.lsk$save_betadiv(marmoset$microeco$beta$fpt$lsk)
fpt.data.rap$save_betadiv(marmoset$microeco$beta$fpt$rap)
fpt.data.rap.g1$save_betadiv(marmoset$microeco$beta$fpt$rap_g1)
fpt.data.rap.g2$save_betadiv(marmoset$microeco$beta$fpt$rap_g2)
fpt.data.rap.int$save_betadiv(marmoset$microeco$beta$fpt$rap_int)
```
### NJC19
```{r}
njc.main        <- trans_func$new(dataset.main)
njc.g1          <- trans_func$new(dataset.g1)
njc.g2          <- trans_func$new(dataset.g2)
njc.int         <- trans_func$new(dataset.int)
njc.lsk <- trans_func$new(dataset.lsk)
njc.rap <- trans_func$new(dataset.rap)
njc.rap.g1 <- trans_func$new(dataset.rap.g1)
njc.rap.g2 <- trans_func$new(dataset.rap.g2)
njc.rap.int <- trans_func$new(dataset.rap.int)

njc.main$cal_spe_func(prok_database = "NJC19")
njc.g1$cal_spe_func(prok_database = "NJC19")
njc.g2$cal_spe_func(prok_database = "NJC19")
njc.int$cal_spe_func(prok_database = "NJC19")
njc.lsk$cal_spe_func(prok_database = "NJC19")
njc.rap$cal_spe_func(prok_database = "NJC19")
njc.rap.g1$cal_spe_func(prok_database = "NJC19")
njc.rap.g2$cal_spe_func(prok_database = "NJC19")
njc.rap.int$cal_spe_func(prok_database = "NJC19")

njc.main$cal_spe_func_perc(abundance_weighted = T)
njc.g1$cal_spe_func_perc(abundance_weighted = T)
njc.g2$cal_spe_func_perc(abundance_weighted = T)
njc.int$cal_spe_func_perc(abundance_weighted = T)
njc.lsk$cal_spe_func_perc(abundance_weighted = T)
njc.rap$cal_spe_func_perc(abundance_weighted = T)
njc.rap.g1$cal_spe_func_perc(abundance_weighted = T)
njc.rap.g2$cal_spe_func_perc(abundance_weighted = T)
njc.rap.int$cal_spe_func_perc(abundance_weighted = T)
```
```{r}
njc.main$trans_spe_func_perc()
njc.g1$trans_spe_func_perc()
njc.g2$trans_spe_func_perc()
njc.int$trans_spe_func_perc()
njc.lsk$trans_spe_func_perc()
njc.rap$trans_spe_func_perc()
njc.rap.g1$trans_spe_func_perc()
njc.rap.g2$trans_spe_func_perc()
njc.rap.int$trans_spe_func_perc()

njc.main.df        <- as.data.frame(t(njc.main$res_spe_func_perc), check.names = F)
njc.g1.df          <- as.data.frame(t(njc.g1$res_spe_func_perc), check.names = F)
njc.g2.df          <- as.data.frame(t(njc.g2$res_spe_func_perc), check.names = F)
njc.int.df         <- as.data.frame(t(njc.int$res_spe_func_perc), check.names = F)
njc.lsk.df <- as.data.frame(t(njc.lsk$res_spe_func_perc), check.names = F)
njc.rap.df <- as.data.frame(t(njc.rap$res_spe_func_perc), check.names = F)
njc.rap.g1.df <- as.data.frame(t(njc.rap.g1$res_spe_func_perc), check.names = F)
njc.rap.g2.df <- as.data.frame(t(njc.rap.g2$res_spe_func_perc), check.names = F)
njc.rap.int.df <- as.data.frame(t(njc.rap.int$res_spe_func_perc), check.names = F)

njc.main.tax        <- as.data.frame(rownames(njc.main.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.g1.tax          <- as.data.frame(rownames(njc.g1.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.g2.tax          <- as.data.frame(rownames(njc.g2.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.int.tax         <- as.data.frame(rownames(njc.int.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.lsk.tax <- as.data.frame(rownames(njc.lsk.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.rap.tax <- as.data.frame(rownames(njc.rap.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.rap.g1.tax <- as.data.frame(rownames(njc.rap.g1.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.rap.g2.tax <- as.data.frame(rownames(njc.rap.g2.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.rap.int.tax <- as.data.frame(rownames(njc.rap.int.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))

njc.main.otu        <- njc.main.df %>% rownames_to_column("Function") %>% left_join(njc.main.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.g1.otu          <- njc.g1.df %>% rownames_to_column("Function") %>% left_join(njc.g1.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.g2.otu          <- njc.g2.df %>% rownames_to_column("Function") %>% left_join(njc.g2.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.int.otu         <- njc.int.df %>% rownames_to_column("Function") %>% left_join(njc.int.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.lsk.otu <- njc.lsk.df %>% rownames_to_column("Function") %>% left_join(njc.lsk.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.rap.otu <- njc.rap.df %>% rownames_to_column("Function") %>% left_join(njc.rap.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.rap.g1.otu <- njc.rap.g1.df %>% rownames_to_column("Function") %>% left_join(njc.rap.g1.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.rap.g2.otu <- njc.rap.g2.df %>% rownames_to_column("Function") %>% left_join(njc.rap.g2.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.rap.int.otu <- njc.rap.int.df %>% rownames_to_column("Function") %>% left_join(njc.rap.int.tax) %>% column_to_rownames("func_name") %>% select(-Function)

njc.main.tax        <- njc.main.tax %>% column_to_rownames("func_name")
njc.g1.tax          <- njc.g1.tax %>% column_to_rownames("func_name")
njc.g2.tax          <- njc.g2.tax %>% column_to_rownames("func_name")
njc.int.tax         <- njc.int.tax %>% column_to_rownames("func_name")
njc.lsk.tax <- njc.lsk.tax %>% column_to_rownames("func_name")
njc.rap.tax <- njc.rap.tax %>% column_to_rownames("func_name")
njc.rap.g1.tax <- njc.rap.g1.tax %>% column_to_rownames("func_name")
njc.rap.g2.tax <- njc.rap.g2.tax %>% column_to_rownames("func_name")
njc.rap.int.tax <- njc.rap.int.tax %>% column_to_rownames("func_name")
```
```{r}
njc.data.main <- microtable$new(otu_table    = njc.main.otu, 
                                tax_table    = njc.main.tax, 
                                sample_table = dataset.main$sample_table)

njc.data.g1 <- microtable$new(otu_table    = njc.g1.otu, 
                                tax_table    = njc.g1.tax, 
                                sample_table = dataset.g1$sample_table)

njc.data.g2 <- microtable$new(otu_table    = njc.g2.otu, 
                                tax_table    = njc.g2.tax, 
                                sample_table = dataset.g2$sample_table)

njc.data.int <- microtable$new(otu_table    = njc.int.otu, 
                                tax_table    = njc.int.tax, 
                                sample_table = dataset.int$sample_table)

njc.data.lsk <- microtable$new(otu_table    = njc.lsk.otu, 
                                tax_table    = njc.lsk.tax, 
                                sample_table = dataset.lsk$sample_table)

njc.data.rap <- microtable$new(otu_table    = njc.rap.otu, 
                                tax_table    = njc.rap.tax, 
                                sample_table = dataset.rap$sample_table)

njc.data.rap.g1 <- microtable$new(otu_table    = njc.rap.g1.otu, 
                                tax_table    = njc.rap.g1.tax, 
                                sample_table = dataset.rap.g1$sample_table)

njc.data.rap.g2 <- microtable$new(otu_table    = njc.rap.g2.otu, 
                                tax_table    = njc.rap.g2.tax, 
                                sample_table = dataset.rap.g2$sample_table)

njc.data.rap.int <- microtable$new(otu_table    = njc.rap.int.otu, 
                                tax_table    = njc.rap.int.tax, 
                                sample_table = dataset.rap.int$sample_table)

njc.data.main$tidy_dataset()
njc.data.g1$tidy_dataset()
njc.data.g2$tidy_dataset()
njc.data.int$tidy_dataset()
njc.data.lsk$tidy_dataset()
njc.data.rap$tidy_dataset()
njc.data.rap.g1$tidy_dataset()
njc.data.rap.g2$tidy_dataset()
njc.data.rap.int$tidy_dataset()

njc.data.main$cal_abund()
njc.data.g1$cal_abund()
njc.data.g2$cal_abund()
njc.data.int$cal_abund()
njc.data.lsk$cal_abund()
njc.data.rap$cal_abund()
njc.data.rap.g1$cal_abund()
njc.data.rap.g2$cal_abund()
njc.data.rap.int$cal_abund()

njc.data.main$cal_alphadiv()
njc.data.g1$cal_alphadiv()
njc.data.g2$cal_alphadiv()
njc.data.int$cal_alphadiv()
njc.data.lsk$cal_alphadiv()
njc.data.rap$cal_alphadiv()
njc.data.rap.g1$cal_alphadiv()
njc.data.rap.g2$cal_alphadiv()
njc.data.rap.int$cal_alphadiv()

njc.data.main$cal_betadiv()
njc.data.g1$cal_betadiv()
njc.data.g2$cal_betadiv()
njc.data.int$cal_betadiv()
njc.data.lsk$cal_betadiv()
njc.data.rap$cal_betadiv()
njc.data.rap.g1$cal_betadiv()
njc.data.rap.g2$cal_betadiv()
njc.data.rap.int$cal_betadiv()
```
```{r}
njc.data.main$save_table(marmoset$microeco$data$njc$main, sep = "\t")
njc.data.g1$save_table(marmoset$microeco$data$njc$g1, sep = "\t")
njc.data.g2$save_table(marmoset$microeco$data$njc$g2, sep = "\t")
njc.data.int$save_table(marmoset$microeco$data$njc$int, sep = "\t")
njc.data.lsk$save_table(marmoset$microeco$data$njc$lsk, sep = "\t")
njc.data.rap$save_table(marmoset$microeco$data$njc$rap, sep = "\t")
njc.data.rap.g1$save_table(marmoset$microeco$data$njc$rap_g1, sep = "\t")
njc.data.rap.g2$save_table(marmoset$microeco$data$njc$rap_g2, sep = "\t")
njc.data.rap.int$save_table(marmoset$microeco$data$njc$rap_int, sep = "\t")

njc.data.main$save_abund(marmoset$microeco$abund$njc$main, sep = "\t")
njc.data.g1$save_abund(marmoset$microeco$abund$njc$g1, sep = "\t")
njc.data.g2$save_abund(marmoset$microeco$abund$njc$g2, sep = "\t")
njc.data.int$save_abund(marmoset$microeco$abund$njc$int, sep = "\t")
njc.data.lsk$save_abund(marmoset$microeco$abund$njc$lsk, sep = "\t")
njc.data.rap$save_abund(marmoset$microeco$abund$njc$rap, sep = "\t")
njc.data.rap.g1$save_abund(marmoset$microeco$abund$njc$rap_g1, sep = "\t")
njc.data.rap.g2$save_abund(marmoset$microeco$abund$njc$rap_g2, sep = "\t")
njc.data.rap.int$save_abund(marmoset$microeco$abund$njc$rap_int, sep = "\t")

njc.data.main$save_alphadiv(marmoset$microeco$alpha$njc$main)
njc.data.g1$save_alphadiv(marmoset$microeco$alpha$njc$g1)
njc.data.g2$save_alphadiv(marmoset$microeco$alpha$njc$g2)
njc.data.int$save_alphadiv(marmoset$microeco$alpha$njc$int)
njc.data.lsk$save_alphadiv(marmoset$microeco$alpha$njc$lsk)
njc.data.rap$save_alphadiv(marmoset$microeco$alpha$njc$rap)
njc.data.rap.g1$save_alphadiv(marmoset$microeco$alpha$njc$rap_g1)
njc.data.rap.g2$save_alphadiv(marmoset$microeco$alpha$njc$rap_g2)
njc.data.rap.int$save_alphadiv(marmoset$microeco$alpha$njc$rap_int)

njc.data.main$save_betadiv(marmoset$microeco$beta$njc$main)
njc.data.g1$save_betadiv(marmoset$microeco$beta$njc$g1)
njc.data.g2$save_betadiv(marmoset$microeco$beta$njc$g2)
njc.data.int$save_betadiv(marmoset$microeco$beta$njc$int)
njc.data.lsk$save_betadiv(marmoset$microeco$beta$njc$lsk)
njc.data.rap$save_betadiv(marmoset$microeco$beta$njc$rap)
njc.data.rap.g1$save_betadiv(marmoset$microeco$beta$njc$rap_g1)
njc.data.rap.g2$save_betadiv(marmoset$microeco$beta$njc$rap_g2)
njc.data.rap.int$save_betadiv(marmoset$microeco$beta$njc$rap_int)
```