---
title: "Marm_MicroecoPrep_2025-5-20_SA"
author: "Shayda"
date: "`r Sys.Date()`"
params:

  
---
Chunk errors = knitting error

```{r}
global <- config::get(config = "default")

here::i_am("Marm_MicroecoPrep_2025-5-20_SA.Rmd")
source(here::here(global$setup))

```
# Metadata
```{r}
subjects <- read.table(marmoset$subjects,
                      header = T,
                      sep = "\t") %>%
  mutate(
    Room = str_glue("Room{str_extract(Cage, '^[a-zA-Z]')}"),
    Subject_Code = factor(Subject_Code, levels = unique(Subject_Code)),
    Subject = factor(Subject, levels = unique(Subject)),
    Sex = factor(Sex, levels = unique(Sex)),
    Housing = factor(Housing, levels = unique(Housing)),
    Group = factor(Group, levels = unique(Group)),
    Subj_DOB = as.Date(DOB, format = "%m/%d/%y"),
    Room = factor(Room, levels = unique(Room))
  ) %>%
  dplyr::select(Group,
                Room,
                Cage,
                Subject,
                Subject_Code,
                Sex,
                Housing,
                Origin,
                Subj_DOB,
                Age,
                Contraception,
                Previous_Exp,
                Subject_Notes
                )
```
```{r}
write.csv(subjects, "metadata/subjects.csv")
```
```{r}
saveRDS(subjects, file = "data/subjects.rds")
```
```{r}
samples <- read.table(marmoset$samples,
                      header = T,
                      sep = "\t") %>%
  mutate(Date = as.Date(Sample_Date,
                        format = "%m/%d/%y"),
         Study_Day = str_extract(Sample_ID, "(?<=-)[0-9]+(?=-)"),
         Subj_Day = fct(str_glue("{Subject}", "_", "{Study_Day}")),
         SampleTimeFormatted = str_pad(Sample_Time, 4, pad = "0"),
         Sample_Time = as_hms(sprintf("%02d:%02d:00",
          as.integer(str_sub(SampleTimeFormatted, 1, 2)),
          as.integer(str_sub(SampleTimeFormatted, 3, 4))))) %>%
  dplyr::select(Sample_ID,
                Date,
                Study_Day,
                Sample_Time,
                Subject,
                Sample_Collected_By,
                Bristol,
                Sample_Stabilizer,
                Sample_Box,
                Sample_Notes,
                Diet_Trial,
                Group_Phase,
                Subj_Day
                )
```
```{r}
write.csv(samples, "metadata/samples.csv")
```
```{r}
saveRDS(samples, file = "data/samples.rds")
```
```{r}
extracts <- read.table(marmoset$extracts,
                      header = T,
                      sep = "\t") %>%
  mutate(Extract_Date = as.Date(Extract_Date,
                               format = "%m/%d/%y"),
         Extract_Conc_ng_uL = Extract_Conc_.ng.uL.) %>%
  dplyr::select(Extract_ID,
                Sample_ID,
                Extract_Type,
                Extract_Date,
                Extracted_By,
                Extract_Conc_ng_uL,
                Extract_Kit,
                Extract_Box,
                Extract_Notes
                )
```
```{r}
write.csv(extracts, "metadata/extracts.csv")
```
```{r}
saveRDS(extracts, file = "data/extracts.rds")
```
```{r}
libraries <- read.table(marmoset$libraries,
                      header = T,
                      sep = "\t") %>%
  mutate(Final_Lib_Conc_ng_uL = Final_Lib_Conc_.ng.uL.,
         Vol_Added_Pool_uL = Vol_Added_Pool_.uL.) %>%
  dplyr::select(Sequence_ID,
                Extract_ID,
                Sample_ID,
                Seq_ID,
                Pipeline,
                Library_Tube,
                Library_Tube_ID,
                Library_Barcode,
                F_Primer,
                R_Primer,
                Run_ID,
                Final_Lib_Conc_ng_uL,
                Vol_Added_Pool_uL
                )
```
```{r}
write.csv(libraries, "metadata/libraries.csv")
```
```{r}
saveRDS(libraries, file = "data/libraries.rds")
```
```{r}
seqruns <- read.table(marmoset$seqruns,
                      header = T,
                      sep = "\t") %>%
  mutate(Run_Date = as.Date(Run_Date,
                               format = "%m/%d/%y"),
         Seq_ID = Pooled_Library_Code,
         Full_Seq_ID = Pooled_Library_ID) %>%
  dplyr::select(Run_ID,
                Full_Seq_ID,
                Seq_ID,
                Pipeline,
                Kit,
                Sequencer,
                MinION_Position,
                Flow_Cell_Type,
                Flongle_Adapter,
                Flow_Cell_ID,
                Run_Date,
                Loaded_By
                )
```
```{r}
write.csv(seqruns, "metadata/seqruns.csv")
```
```{r}
saveRDS(seqruns, file = "data/seqruns.rds")
```
```{r}
weights <- read.table(marmoset$weights,
                      header = T,
                      sep = "\t") %>%
  mutate(Date = as.Date(Date,
                        format = "%m/%d/%y")) %>%
  dplyr::select(Date,
                Study_Day,
                Subject,
                Weight_g,
                Weight_Collected_By,
                Weight_Notes)
```
```{r}
write.csv(weights, "metadata/weights.csv")
```
```{r}
saveRDS(weights, file = "data/weights.rds")
```
```{r}
leftover_food <- read.table(marmoset$leftover_food,
                      header = T,
                      sep = "\t") %>%
  mutate(Date = as.Date(Date,
                                format = "%m/%d/%y"),
         Leftover_Gel_Zupreem = Gel.Zupreem,
         Leftover_Apple = Apple,
         Leftover_Sweet_Potato = Sweet_Potato,
         Leftover_Eggs = Eggs) %>%
  dplyr::select(Date,
                Cage,
                Leftover_Gel_Zupreem,
                Leftover_Apple,
                Leftover_Sweet_Potato,
                Leftover_Eggs,
                Diet_Notes)
```
```{r}
write.csv(leftover_food, "metadata/leftover_food.csv")
```
```{r}
saveRDS(leftover_food, file = "data/leftover_food.rds")
```
```{r}
bristol <- read.table(marmoset$bristol,
                      header = T,
                      sep = "\t") %>%
  mutate(Study_Date = as.Date(Date,
                                format = "%m/%d/%y"),
         SampleTimeFormatted = str_pad(Time, 4, pad = "0"),
         Sample_Time = as_hms(sprintf("%02d:%02d:00",
          as.integer(str_sub(SampleTimeFormatted, 1, 2)),
          as.integer(str_sub(SampleTimeFormatted, 3, 4))))) %>%
  dplyr::select(Study_Date,
                Study_Day,
                Sample_Time,
                Subject,
                Bristol
                )
```
```{r}
write.csv(bristol, "metadata/bristol.csv")
```
```{r}
saveRDS(bristol, file = "data/bristol.rds")
```
```{r}
seq_metadata <- libraries %>%
  left_join(seqruns, by = c("Seq_ID", "Run_ID")) %>%
  dplyr::select(-"Pipeline.x") %>%
  mutate(Pipeline = Pipeline.y) %>%
  dplyr::select(-"Pipeline.y")
```
```{r}
metadata <- seq_metadata %>%
  left_join(extracts, by = c("Extract_ID", "Sample_ID")) %>%
  left_join(samples, by = "Sample_ID") %>%
  left_join(subjects, by = "Subject")
```
```{r}
write.csv(metadata, "metadata/metadata.csv")
```
```{r}
saveRDS(metadata, file = "data/metadata.rds")
```

## Sample Subsets
```{r}
sample.list.g1 <- metadata %>%
  filter(Group == "Group1") %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.g1 <- sample.list.g1 %>% unlist()

sample.list.g2 <- metadata %>%
  filter(Group == "Group2") %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.g2 <- sample.list.g2 %>% unlist()

sample.list.int <- metadata %>%
  filter(Group_Phase == c("gum", "control")) %>%
  distinct(Sequence_ID, Study_Day) %>%
  arrange(Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

samples.int <- sample.list.int %>% unlist()
```
# Alignment
*Generated from the wf-16s workflow*
```{r}
alignment.files.list <- list.files(path = marmoset$read_alignments,
                                   pattern = "*_wf-metagenomics-alignment.csv$",
                                   full.names = TRUE)

alignment.filenames <- list.files(path = marmoset$read_alignments,
                                   pattern = "*_wf-metagenomics-alignment.csv$",
                                   full.names = FALSE)

alignment.files <- lapply(alignment.files.list,
                          read_alignment_file)

alias <- str_remove(alignment.filenames, 
                    "_wf-metagenomics-alignment.csv")
```
```{r}
alignment.files <- Map(function(df, id) {
  if (nrow(df) > 0) {
    df$alias <- id
  } else {
    warning(paste("Data frame for", id, "is empty. Alias not assigned."))
  }
  return(df)
}, alignment.files, alias)
```
*Filter reads based on a minimum coverage of 80*
```{r}
alignments.long <- bind_rows(alignment.files) %>%
                        as_tibble() %>% fix.strings() -> alignments.clean

alignments.long <- dplyr::select(alignments.clean,
                                reference,
                               taxid,
                               species,
                               genus,
                               family,
                               order,
                               class,
                               phylum,
                               superkingdom,
                               Sequence_ID = alias,
                               pc.coverage,
                               n_reads = number.of.reads) %>%
 left_join(dplyr::select(metadata, Sequence_ID), 
            by = join_by(Sequence_ID)) %>%
  dplyr::filter(pc.coverage >= methods_16s$min_cov & !is.na(Sequence_ID)) %>%
  group_by(reference,
           taxid,
           species,
           genus,
           family,
           order,
           class,
           phylum,
           superkingdom,
           Sequence_ID) %>%
  summarize(samp_cov = mean(pc.coverage),
            samp_n_reads = mean(n_reads)) %>% ungroup() %>%
  arrange(Sequence_ID) %>%
  sort.taxa()
```
*Simplify into a table with one row per reference*
```{r}
alignment.refs <- alignments.long %>%
  group_by(reference,
           taxid,
           species,
           genus,
           family,
           order,
           class,
           phylum,
           superkingdom) %>%
  summarize(ref_n_reads = round(sum(samp_n_reads), digits = 2),
            ref_mean_cov = round(mean(samp_cov), digits = 2)) %>% ungroup() %>%
  group_by(species,
           genus,
           family,
           order,
           class,
           phylum,
           superkingdom) %>%
  arrange(species, desc(ref_n_reads), desc(ref_mean_cov)) %>%
                        mutate(ref_order       = row_number(),
                               tax_total_count = sum(ref_n_reads)) %>%
                        ungroup() %>%
                        filter(ref_order == 1) %>%
                        select(-ref_order) %>% sort.taxa()
```
*Format alignment table structure for calculation of relative abundances*
```{r}
alignments    <- alignments.long %>% 
                        group_by(superkingdom,
                                 phylum,
                                 class,
                                 order,
                                 family,
                                 genus,
                                 species,
                                 Sequence_ID) %>%
                          summarize(counts = sum(samp_n_reads)) %>%
                          ungroup() %>%
                          left_join(alignment.refs,
                    by = join_by(superkingdom,
                                 phylum,
                                 class,
                                 order,
                                 family,
                                 genus,
                                 species)) %>%
                    pivot_wider(id_cols    = c(reference,
                                               taxid,
                                               all_of(taxonomy.ordered),
                                               ref_mean_cov),
                                names_from  = Sequence_ID,
                                values_from = counts) %>%
                    sort.taxa() %>%
                    mutate(across(where(is.numeric), ~replace_na(.x, 0)),
                           organism = str_glue("txid{taxid}"))
```
## Sequencing Depth
```{r}
depth <- alignments.long %>%
  filter(!is.na(Sequence_ID)) %>%
  select(Sequence_ID,
         samp_cov,
         samp_n_reads) %>%
  group_by(Sequence_ID) %>%
  summarize(depth         = round(sum(samp_n_reads), digits = 0),
            mean_coverage = round(mean(samp_cov), digits = 2)) %>%
  mutate(mean_coverage = mean_coverage/100) %>%
  ungroup() %>%
  left_join(metadata, by = join_by(Sequence_ID))
```
## Reference Sequences
*Generate an accession reference table for sequences to retrieve*
```{r}
new.refs          <- alignments %>% 
                     select(reference, organism) %>% distinct()

  write.table(new.refs,
              global$tmp_fetch,
              sep       = "\t",
              quote     = FALSE,
              row.names = FALSE,
              col.names = FALSE)
```
*(->Terminal: fetch_refs_cm)*
## FASTA
*Generate fasta to align sequences and assemble phylo tree*
```{r}
new.refseqs <- read.fasta(paste0(global$tmp_fasta3))

write.fasta(sequences = new.refseqs,
            names = names(new.refseqs),
            file.out = global$tmp_fasta4)
```
*(->Terminal: align_tree_cm)*
## Phylogenetic Tree and Representative Fasta
```{r}
tax.tree <- read.tree(marmoset$refseq_tree)
rep.seqs <- read.fasta(marmoset$refseq_fasta)
rep.seqs.tax4fun <- lapply(rep.seqs, function(seq) {gsub("-", "", seq)})
```
# Prepare MicroEco Datasets
## Taxonomy
```{r}
tax.table <- alignments %>%
  column_to_rownames("organism") %>%
  select(all_of(taxonomy.ordered)) %>%
  sort.taxa() %>%
  rename_with(~str_to_title(.x), .cols = everything()) %>%
             mutate(Kingdom = Superkingdom, .keep = "unused") %>%
             relocate(Kingdom) %>%
             distinct() %>%
             tidy_taxonomy()
```
```{r}
taxonomy.list <- alignments %>% 
                 select(organism, taxonomy.ordered) %>% 
                 sort.taxa() %>%
                 rename_with(~str_to_title(.x), .cols = everything()) %>%
                 mutate(Kingdom = Superkingdom, .keep = "unused") %>%
                 relocate(Organism, Kingdom)

write.table(taxonomy.list,
            marmoset$taxonomy_list,
            row.names = FALSE,
            sep = "\t")
```
## OTUs
```{r}
sample.list <- metadata %>%
  distinct(Sequence_ID, Subject, Study_Day) %>%
  arrange(Subject, Study_Day) %>%
  distinct(Sequence_ID) %>%
  map(\(x) as.list(x)) %>%
  list_flatten(name_spec = "")

seqsamples <- sample.list %>% unlist()
```
```{r}
otu.table <- alignments %>% 
          select(organism,
                 any_of(seqsamples)) %>%
                 column_to_rownames("organism")

otu.sample.list <- as.list(names(otu.table))
```
## Sample Table
```{r}
samples.filtered <- setdiff(sample.list, otu.sample.list)

excluded <- metadata %>%
  filter(Sequence_ID %in% samples.filtered)
```
```{r}
sample.table <- metadata %>%
  filter(Sequence_ID %in% otu.sample.list) %>%
  rename_with(~str_replace_all(., " ", "_")) %>%
                rename_with(~str_replace_all(., "-", "_")) %>%
                rename_with(~str_remove_all(., fixed(")"))) %>%
                rename_with(~str_remove_all(., fixed("("))) %>%
                arrange(Study_Day, Subject) %>%
                mutate(day = as.character(Study_Day)) %>%
  column_to_rownames("Sequence_ID") 

write.table(sample.table, marmoset$sample_table)
```
### Merged Sample Table
```{r}
sample.table.merged <- sample.table %>%
  remove_rownames() %>%
  select(all_of(sample_merge_cols)) %>%
  distinct() %>%
  filter(!is.na(Subj_Day)) %>%
  column_to_rownames("Subj_Day") 

write.table(sample.table.merged, marmoset$sample_table)
  
```
## Data Subsets
```{r}
otu.g1 <- otu.table %>%
  select(any_of(samples.g1))
otu.g2 <- otu.table %>%
  select(any_of(samples.g2))


otu.int <- otu.table %>%
  select(any_of(samples.int))



samp.table.g1 <- sample.table %>%
  filter(group == "g1")
samp.table.g2 <- sample.table %>%
  filter(group == "g2")


samp.table.int <- sample.table %>%
  filter(group_phase == c("gum", "control"))


```
# Assemble MicroEco Datasets
```{r}
dataset.main <- microtable$new(
  sample_table = sample.table,
  otu_table = otu.table,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.g1 <- microtable$new(
  sample_table = samp.table.g1,
  otu_table = otu.g1,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.g2 <- microtable$new(
  sample_table = samp.table.g2,
  otu_table = otu.g2,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)

dataset.int <- microtable$new(
  sample_table = samp.table.int,
  otu_table = otu.int,
  tax_table = tax.table,
  phylo_tree = tax.tree,
  rep_fasta = rep.seqs,
  auto_tidy = T)
```
# MicroEco Basic Stats and Data Cleaning
## Rarefaction
*Generate rarefaction plot to determine value*
```{r}
rarefaction_depths <- c(seq.int(0, 10000, by = 500))

rarefaction.main <- trans_rarefy$new(dataset.main,
                                     alphadiv = "Observed",
                                     depth = rarefaction_depths)
rarefaction_plot <- rarefaction.main$plot_rarefy(show_legend = FALSE)
rarefaction_plot
```
```{r}
ggsave("visuals/marm_rarefaction.png", rarefaction_plot)
```
*Rarefy samples (norm = "SRS", sample.size = 4000)*
```{r}
dataset.main$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.g1$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.g2$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.int$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)

```
## Technical Replicates
*Merge samples for duplicate observations*
```{r}
dataset.main <- dataset.main$merge_samples("subj_day")
dataset.g1 <- dataset.g1$merge_samples("subj_day")
dataset.g2 <- dataset.g2$merge_samples("subj_day")
dataset.int <- dataset.int$merge_samples("subj_day")
```
```{r}
dataset.main$sample_table <- data.frame(sample.table.merged[rownames(dataset.main$sample_table), ])
dataset.g1$sample_table <- data.frame(sample.table.merged[rownames(dataset.g1$sample_table), ])
dataset.g2$sample_table <- data.frame(sample.table.merged[rownames(dataset.g2$sample_table), ])
dataset.int$sample_table <- data.frame(sample.table.merged[rownames(dataset.int$sample_table), ])
```
## Relative Abundance
```{r}
dataset.main$cal_abund()
dataset.g1$cal_abund()
dataset.g2$cal_abund()
dataset.int$cal_abund()

```
### Filter Taxa
*Remove taxa below an abundance of 0.00001 and a frequency of 1*
```{r}
dataset.main$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.g1$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.g2$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)

dataset.int$filter_taxa(rel_abund = methods_16s$marm$min_abund,
                         freq = methods_16s$marm$min_freq,
                         include_lowest = methods_16s$marm$include_lowest)
```
## Diversity Metrics
### Alpha Diversity
*Calculate alpha diversity indices (including Faith's Phylogenetic Diversity)*
```{r}
dataset.main$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.g1$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.g2$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
dataset.int$cal_alphadiv(PD = methods_16s$marm$alpha_pd)
```
### Beta Diversity
*Calculate beta diversity (unifrac = TRUE, method = aitchison)*
```{r}
dataset.main$cal_betadiv(unifrac = methods_16s$marm$unifrac,
                         method = methods_16s$marm$betadiv)
```
```{r}
dataset.g1$cal_betadiv(unifrac = methods_16s$marm$unifrac,
                         method = methods_16s$marm$betadiv)
```
```{r}
dataset.g2$cal_betadiv(unifrac = methods_16s$marm$unifrac,
                         method = methods_16s$marm$betadiv)
```
```{r}
dataset.int$cal_betadiv(unifrac = methods_16s$marm$unifrac,
                         method = methods_16s$marm$betadiv)
```
## Export Main Datasets
```{r}
dataset.main$save_table(marmoset$microeco$data$main, sep = "\t")
dataset.g1$save_table(marmoset$microeco$data$g1, sep = "\t")
dataset.g2$save_table(marmoset$microeco$data$g2, sep = "\t")
dataset.int$save_table(marmoset$microeco$data$int, sep = "\t")

dataset.main$save_abund(marmoset$microeco$abund$main, sep = "\t")
dataset.g1$save_abund(marmoset$microeco$abund$g1, sep = "\t")
dataset.g2$save_abund(marmoset$microeco$abund$g2, sep = "\t")
dataset.int$save_abund(marmoset$microeco$abund$int, sep = "\t")

dataset.main$save_alphadiv(marmoset$microeco$alpha$main)
dataset.g1$save_alphadiv(marmoset$microeco$alpha$g1)
dataset.g2$save_alphadiv(marmoset$microeco$alpha$g2)
dataset.int$save_alphadiv(marmoset$microeco$alpha$int)


dataset.main$save_betadiv(marmoset$microeco$beta$main)
dataset.g1$save_betadiv(marmoset$microeco$beta$g1)
dataset.g2$save_betadiv(marmoset$microeco$beta$g2)
dataset.int$save_betadiv(marmoset$microeco$beta$int)
```
## Taxonomic Level Subsets
*Create dataframe subsets for each taxonomic level*
### Phylum
```{r}
phy.data.main <- clone(dataset.main, deep = T)
phy.data.g1 <- clone(dataset.g1, deep = T)
phy.data.g2 <- clone(dataset.g2, deep = T)
phy.data.int <- clone(dataset.int, deep = T)

phy.main <- phy.data.main$merge_taxa(taxa = "Phylum")
phy.main$tax_table %<>% .[.$Phylum != "p__", ]
phy.main$tidy_dataset()
rownames(phy.main$otu_table) <- phy.main$tax_table[rownames(phy.main$otu_table), "Phylum"]
rownames(phy.main$tax_table) <- phy.main$tax_table[, "Phylum"]

phy.g1 <- phy.data.g1$merge_taxa(taxa = "Phylum")
phy.g1$tax_table %<>% .[.$Phylum != "p__", ]
phy.g1$tidy_dataset()
rownames(phy.g1$otu_table) <- phy.g1$tax_table[rownames(phy.g1$otu_table), "Phylum"]
rownames(phy.g1$tax_table) <- phy.g1$tax_table[, "Phylum"]

phy.g2 <- phy.data.g2$merge_taxa(taxa = "Phylum")
phy.g2$tax_table %<>% .[.$Phylum != "p__", ]
phy.g2$tidy_dataset()
rownames(phy.g2$otu_table) <- phy.g2$tax_table[rownames(phy.g2$otu_table), "Phylum"]
rownames(phy.g2$tax_table) <- phy.g2$tax_table[, "Phylum"]

phy.int <- phy.data.int$merge_taxa(taxa = "Phylum")
phy.int$tax_table %<>% .[.$Phylum != "p__", ]
phy.int$tidy_dataset()
rownames(phy.int$otu_table) <- phy.int$tax_table[rownames(phy.int$otu_table), "Phylum"]
rownames(phy.int$tax_table) <- phy.int$tax_table[, "Phylum"]
```
### Class
```{r}
cla.data.main <- clone(dataset.main, deep = T)
cla.data.g1 <- clone(dataset.g1, deep = T)
cla.data.g2 <- clone(dataset.g2, deep = T)
cla.data.int <- clone(dataset.int, deep = T)

cla.main <- cla.data.main$merge_taxa(taxa = "Class")
cla.main$tax_table %<>% .[.$Class != "c__", ]
cla.main$tidy_dataset()
rownames(cla.main$otu_table) <- cla.main$tax_table[rownames(cla.main$otu_table), "Class"]
rownames(cla.main$tax_table) <- cla.main$tax_table[, "Class"]

cla.g1 <- cla.data.g1$merge_taxa(taxa = "Class")
cla.g1$tax_table %<>% .[.$Class != "c__", ]
cla.g1$tidy_dataset()
rownames(cla.g1$otu_table) <- cla.g1$tax_table[rownames(cla.g1$otu_table), "Class"]
rownames(cla.g1$tax_table) <- cla.g1$tax_table[, "Class"]

cla.g2 <- cla.data.g2$merge_taxa(taxa = "Class")
cla.g2$tax_table %<>% .[.$Class != "c__", ]
cla.g2$tidy_dataset()
rownames(cla.g2$otu_table) <- cla.g2$tax_table[rownames(cla.g2$otu_table), "Class"]
rownames(cla.g2$tax_table) <- cla.g2$tax_table[, "Class"]

cla.int <- cla.data.int$merge_taxa(taxa = "Class")
cla.int$tax_table %<>% .[.$Class != "c__", ]
cla.int$tidy_dataset()
rownames(cla.int$otu_table) <- cla.int$tax_table[rownames(cla.int$otu_table), "Class"]
rownames(cla.int$tax_table) <- cla.int$tax_table[, "Class"]
```
### Order
```{r}
ord.data.main <- clone(dataset.main, deep = T)
ord.data.g1 <- clone(dataset.g1, deep = T)
ord.data.g2 <- clone(dataset.g2, deep = T)
ord.data.int <- clone(dataset.int, deep = T)

ord.main <- ord.data.main$merge_taxa(taxa = "Order")
ord.main$tax_table %<>% .[.$Order != "o__", ]
ord.main$tidy_dataset()
rownames(ord.main$otu_table) <- ord.main$tax_table[rownames(ord.main$otu_table), "Order"]
rownames(ord.main$tax_table) <- ord.main$tax_table[, "Order"]

ord.g1 <- ord.data.g1$merge_taxa(taxa = "Order")
ord.g1$tax_table %<>% .[.$Order != "o__", ]
ord.g1$tidy_dataset()
rownames(ord.g1$otu_table) <- ord.g1$tax_table[rownames(ord.g1$otu_table), "Order"]
rownames(ord.g1$tax_table) <- ord.g1$tax_table[, "Order"]

ord.g2 <- ord.data.g2$merge_taxa(taxa = "Order")
ord.g2$tax_table %<>% .[.$Order != "o__", ]
ord.g2$tidy_dataset()
rownames(ord.g2$otu_table) <- ord.g2$tax_table[rownames(ord.g2$otu_table), "Order"]
rownames(ord.g2$tax_table) <- ord.g2$tax_table[, "Order"]

ord.int <- ord.data.int$merge_taxa(taxa = "Order")
ord.int$tax_table %<>% .[.$Order != "o__", ]
ord.int$tidy_dataset()
rownames(ord.int$otu_table) <- ord.int$tax_table[rownames(ord.int$otu_table), "Order"]
rownames(ord.int$tax_table) <- ord.int$tax_table[, "Order"]
```
### Family
```{r}
fam.data.main <- clone(dataset.main, deep = T)
fam.data.g1 <- clone(dataset.g1, deep = T)
fam.data.g2 <- clone(dataset.g2, deep = T)
fam.data.int <- clone(dataset.int, deep = T)

fam.main <- fam.data.main$merge_taxa(taxa = "Family")
fam.main$tax_table %<>% .[.$Family != "f__", ]
fam.main$tidy_dataset()
rownames(fam.main$otu_table) <- fam.main$tax_table[rownames(fam.main$otu_table), "Family"]
rownames(fam.main$tax_table) <- fam.main$tax_table[, "Family"]

fam.g1 <- fam.data.g1$merge_taxa(taxa = "Family")
fam.g1$tax_table %<>% .[.$Family != "f__", ]
fam.g1$tidy_dataset()
rownames(fam.g1$otu_table) <- fam.g1$tax_table[rownames(fam.g1$otu_table), "Family"]
rownames(fam.g1$tax_table) <- fam.g1$tax_table[, "Family"]

fam.g2 <- fam.data.g2$merge_taxa(taxa = "Family")
fam.g2$tax_table %<>% .[.$Family != "f__", ]
fam.g2$tidy_dataset()
rownames(fam.g2$otu_table) <- fam.g2$tax_table[rownames(fam.g2$otu_table), "Family"]
rownames(fam.g2$tax_table) <- fam.g2$tax_table[, "Family"]

fam.int <- fam.data.int$merge_taxa(taxa = "Family")
fam.int$tax_table %<>% .[.$Family != "f__", ]
fam.int$tidy_dataset()
rownames(fam.int$otu_table) <- fam.int$tax_table[rownames(fam.int$otu_table), "Family"]
rownames(fam.int$tax_table) <- fam.int$tax_table[, "Family"]
```
### Genus
```{r}
gen.data.main <- clone(dataset.main, deep = T)
gen.data.g1 <- clone(dataset.g1, deep = T)
gen.data.g2 <- clone(dataset.g2, deep = T)
gen.data.int <- clone(dataset.int, deep = T)

gen.main <- gen.data.main$merge_taxa(taxa = "Genus")
gen.main$tax_table %<>% .[.$Genus != "g__", ]
gen.main$tidy_dataset()
rownames(gen.main$otu_table) <- gen.main$tax_table[rownames(gen.main$otu_table), "Genus"]
rownames(gen.main$tax_table) <- gen.main$tax_table[, "Genus"]

gen.g1 <- gen.data.g1$merge_taxa(taxa = "Genus")
gen.g1$tax_table %<>% .[.$Genus != "g__", ]
gen.g1$tidy_dataset()
rownames(gen.g1$otu_table) <- gen.g1$tax_table[rownames(gen.g1$otu_table), "Genus"]
rownames(gen.g1$tax_table) <- gen.g1$tax_table[, "Genus"]

gen.g2 <- gen.data.g2$merge_taxa(taxa = "Genus")
gen.g2$tax_table %<>% .[.$Genus != "g__", ]
gen.g2$tidy_dataset()
rownames(gen.g2$otu_table) <- gen.g2$tax_table[rownames(gen.g2$otu_table), "Genus"]
rownames(gen.g2$tax_table) <- gen.g2$tax_table[, "Genus"]

gen.int <- gen.data.int$merge_taxa(taxa = "Genus")
gen.int$tax_table %<>% .[.$Genus != "g__", ]
gen.int$tidy_dataset()
rownames(gen.int$otu_table) <- gen.int$tax_table[rownames(gen.int$otu_table), "Genus"]
rownames(gen.int$tax_table) <- gen.int$tax_table[, "Genus"]
```
## Export Taxonomic Subsets
```{r}
phy.main$save_table(marmoset$microeco$data$phylum$main, sep = "\t")
phy.g1$save_table(marmoset$microeco$data$phylum$g1, sep = "\t")
phy.g2$save_table(marmoset$microeco$data$phylum$g2, sep = "\t")
phy.int$save_table(marmoset$microeco$data$phylum$int, sep = "\t")

phy.main$save_abund(marmoset$microeco$abund$phylum$main, sep = "\t")
phy.g1$save_abund(marmoset$microeco$abund$phylum$g1, sep = "\t")
phy.g2$save_abund(marmoset$microeco$abund$phylum$g2, sep = "\t")
phy.int$save_abund(marmoset$microeco$abund$phylum$int, sep = "\t")

phy.main$save_alphadiv(marmoset$microeco$alpha$phylum$main)
phy.g1$save_alphadiv(marmoset$microeco$alpha$phylum$g1)
phy.g2$save_alphadiv(marmoset$microeco$alpha$phylum$g2)
phy.int$save_alphadiv(marmoset$microeco$alpha$phylum$int)

phy.main$save_betadiv(marmoset$microeco$beta$phylum$main)
phy.g1$save_betadiv(marmoset$microeco$beta$phylum$g1)
phy.g2$save_betadiv(marmoset$microeco$beta$phylum$g2)
phy.int$save_betadiv(marmoset$microeco$beta$phylum$int)
```
```{r}
cla.main$save_table(marmoset$microeco$data$class$main, sep = "\t")
cla.g1$save_table(marmoset$microeco$data$class$g1, sep = "\t")
cla.g2$save_table(marmoset$microeco$data$class$g2, sep = "\t")
cla.int$save_table(marmoset$microeco$data$class$int, sep = "\t")

cla.main$save_abund(marmoset$microeco$abund$class$main, sep = "\t")
cla.g1$save_abund(marmoset$microeco$abund$class$g1, sep = "\t")
cla.g2$save_abund(marmoset$microeco$abund$class$g2, sep = "\t")
cla.int$save_abund(marmoset$microeco$abund$class$int, sep = "\t")

cla.main$save_alphadiv(marmoset$microeco$alpha$class$main)
cla.g1$save_alphadiv(marmoset$microeco$alpha$class$g1)
cla.g2$save_alphadiv(marmoset$microeco$alpha$class$g2)
cla.int$save_alphadiv(marmoset$microeco$alpha$class$int)

cla.main$save_betadiv(marmoset$microeco$beta$class$main)
cla.g1$save_betadiv(marmoset$microeco$beta$class$g1)
cla.g2$save_betadiv(marmoset$microeco$beta$class$g2)
cla.int$save_betadiv(marmoset$microeco$beta$class$int)
```
```{r}
ord.main$save_table(marmoset$microeco$data$order$main, sep = "\t")
ord.g1$save_table(marmoset$microeco$data$order$g1, sep = "\t")
ord.g2$save_table(marmoset$microeco$data$order$g2, sep = "\t")
ord.int$save_table(marmoset$microeco$data$order$int, sep = "\t")

ord.main$save_abund(marmoset$microeco$abund$order$main, sep = "\t")
ord.g1$save_abund(marmoset$microeco$abund$order$g1, sep = "\t")
ord.g2$save_abund(marmoset$microeco$abund$order$g2, sep = "\t")
ord.int$save_abund(marmoset$microeco$abund$order$int, sep = "\t")

ord.main$save_alphadiv(marmoset$microeco$alpha$order$main)
ord.g1$save_alphadiv(marmoset$microeco$alpha$order$g1)
ord.g2$save_alphadiv(marmoset$microeco$alpha$order$g2)
ord.int$save_alphadiv(marmoset$microeco$alpha$order$int)

ord.main$save_betadiv(marmoset$microeco$beta$order$main)
ord.g1$save_betadiv(marmoset$microeco$beta$order$g1)
ord.g2$save_betadiv(marmoset$microeco$beta$order$g2)
ord.int$save_betadiv(marmoset$microeco$beta$order$int)
```
```{r}
fam.main$save_table(marmoset$microeco$data$family$main, sep = "\t")
fam.g1$save_table(marmoset$microeco$data$family$g1, sep = "\t")
fam.g2$save_table(marmoset$microeco$data$family$g2, sep = "\t")
fam.int$save_table(marmoset$microeco$data$family$int, sep = "\t")

fam.main$save_abund(marmoset$microeco$abund$family$main, sep = "\t")
fam.g1$save_abund(marmoset$microeco$abund$family$g1, sep = "\t")
fam.g2$save_abund(marmoset$microeco$abund$family$g2, sep = "\t")
fam.int$save_abund(marmoset$microeco$abund$family$int, sep = "\t")

fam.main$save_alphadiv(marmoset$microeco$alpha$family$main)
fam.g1$save_alphadiv(marmoset$microeco$alpha$family$g1)
fam.g2$save_alphadiv(marmoset$microeco$alpha$family$g2)
fam.int$save_alphadiv(marmoset$microeco$alpha$family$int)

fam.main$save_betadiv(marmoset$microeco$beta$family$main)
fam.g1$save_betadiv(marmoset$microeco$beta$family$g1)
fam.g2$save_betadiv(marmoset$microeco$beta$family$g2)
fam.int$save_betadiv(marmoset$microeco$beta$family$int)
```
```{r}
gen.main$save_table(marmoset$microeco$data$genus$main, sep = "\t")
gen.g1$save_table(marmoset$microeco$data$genus$g1, sep = "\t")
gen.g2$save_table(marmoset$microeco$data$genus$g2, sep = "\t")
gen.int$save_table(marmoset$microeco$data$genus$int, sep = "\t")

gen.main$save_abund(marmoset$microeco$abund$genus$main, sep = "\t")
gen.g1$save_abund(marmoset$microeco$abund$genus$g1, sep = "\t")
gen.g2$save_abund(marmoset$microeco$abund$genus$g2, sep = "\t")
gen.int$save_abund(marmoset$microeco$abund$genus$int, sep = "\t")

gen.main$save_alphadiv(marmoset$microeco$alpha$genus$main)
gen.g1$save_alphadiv(marmoset$microeco$alpha$genus$g1)
gen.g2$save_alphadiv(marmoset$microeco$alpha$genus$g2)
gen.int$save_alphadiv(marmoset$microeco$alpha$genus$int)

gen.main$save_betadiv(marmoset$microeco$beta$genus$main)
gen.g1$save_betadiv(marmoset$microeco$beta$genus$g1)
gen.g2$save_betadiv(marmoset$microeco$beta$genus$g2)
gen.int$save_betadiv(marmoset$microeco$beta$genus$int)
```
## Functional Profiles
### KEGG
```{r}
dataset.keg.main <- microtable$new(
                               sample_table = sample.table,
                               otu_table    = otu.table,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.g1 <- microtable$new(
                               sample_table = samp.table.g1,
                               otu_table    = otu.g1,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.g2 <- microtable$new(
                               sample_table = samp.table.g2,
                               otu_table    = otu.g2,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
dataset.keg.int <- microtable$new(
                               sample_table = samp.table.int,
                               otu_table    = otu.int,
                               tax_table    = tax.table,
                               phylo_tree   = tax.tree,
                               rep_fasta    = rep.seqs.tax4fun,
                               auto_tidy    = T)
```
```{r}
dataset.keg.main$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.keg.g1$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.keg.g2$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
dataset.keg.int$rarefy_samples(method = methods_16s$marm$norm, sample.size = methods_16s$marm$rarefy)
```
```{r}
dataset.keg.main <- dataset.keg.main$merge_samples("subj_day")
dataset.keg.g1 <- dataset.keg.g1$merge_samples("subj_day")
dataset.keg.g2 <- dataset.keg.g2$merge_samples("subj_day")
dataset.keg.int <- dataset.keg.int$merge_samples("subj_day")
```
```{r}
dataset.keg.main$sample_table <- data.frame(dataset.keg.main$sample_table, sample.table.merged[rownames(dataset.keg.main$sample_table), ])
dataset.keg.g1$sample_table <- data.frame(dataset.keg.g1$sample_table, sample.table.merged[rownames(dataset.keg.g1$sample_table), ])
dataset.keg.g2$sample_table <- data.frame(dataset.keg.g2$sample_table, sample.table.merged[rownames(dataset.keg.g2$sample_table), ])
dataset.keg.int$sample_table <- data.frame(dataset.keg.int$sample_table, sample.table.merged[rownames(dataset.keg.int$sample_table), ])
```
```{r}
dataset.keg.main$cal_abund()
dataset.keg.g1$cal_abund()
dataset.keg.g2$cal_abund()
dataset.keg.int$cal_abund()
```
```{r}
dataset.keg.main$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.g1$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.g2$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
dataset.keg.int$filter_taxa(
                         rel_abund      = methods_16s$marm$min_abund, 
                         freq           = methods_16s$marm$min_freq, 
                         include_lowest = methods_16s$marm$include_lowest)
```
```{r}
keg.main <- trans_func$new(dataset.keg.main)
keg.g1 <- trans_func$new(dataset.keg.g1)
keg.g2 <- trans_func$new(dataset.keg.g2)
keg.int <- trans_func$new(dataset.keg.int)

```
```{r}
keg.main$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.g1$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.g2$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
keg.int$cal_tax4fun2(
          blast_tool_path           = microbiome$blast,
          path_to_reference_data    = microbiome$tax4fun,
          database_mode             = methods_16s$marm$tax4fun_db, 
          path_to_temp_folder       = microbiome$test_prediction,
          num_threads               = parallel::detectCores() - 1)
```
```{r}
data("Tax4Fun2_KEGG")
```
```{r}
keg.main$cal_tax4fun2_FRI()
```
```{r}
keg.data.main <- microtable$new(
  otu_table = keg.main$res_tax4fun2_pathway,
  tax_table = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.main$sample_table)

keg.data.main$tidy_dataset()

dataset.keg.main$sample_table <- data.frame(dataset.keg.main$sample_table, sample.table.merged[rownames(dataset.keg.main$sample_table), ])

keg.data.main$cal_abund()
keg.data.main$cal_alphadiv()
keg.data.main$cal_betadiv(method = methods_16s$marm$betadiv)
keg.data.main$save_table(marmoset$microeco$data$keg$main, sep = "\t")
keg.data.main$save_abund(marmoset$microeco$abund$keg$main, sep = "\t")
keg.data.main$save_alphadiv(marmoset$microeco$alpha$keg$main)
keg.data.main$save_betadiv(marmoset$microeco$beta$keg$main)
```
```{r}
keg.g1$cal_tax4fun2_FRI()
```
```{r}
keg.data.g1 <- microtable$new(
  otu_table = keg.g1$res_tax4fun2_pathway,
  tax_table = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.g1$sample_table)

keg.data.g1$tidy_dataset()

dataset.keg.g1$sample_table <- data.frame(dataset.keg.g1$sample_table, sample.table.merged[rownames(dataset.keg.g1$sample_table), ])

keg.data.g1$cal_abund()
keg.data.g1$cal_alphadiv()
keg.data.g1$cal_betadiv(method = methods_16s$marm$betadiv)
keg.data.g1$save_table(marmoset$microeco$data$keg$g1, sep = "\t")
keg.data.g1$save_abund(marmoset$microeco$abund$keg$g1, sep = "\t")
keg.data.g1$save_alphadiv(marmoset$microeco$alpha$keg$g1)
keg.data.g1$save_betadiv(marmoset$microeco$beta$keg$g1)
```
```{r}
keg.g2$cal_tax4fun2_FRI()
```
```{r}
keg.data.g2 <- microtable$new(
  otu_table    = keg.g2$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.g2$sample_table)

keg.data.g2$tidy_dataset()

dataset.keg.g2$sample_table <- data.frame(dataset.keg.g2$sample_table, sample.table.merged[rownames(dataset.keg.g2$sample_table), ])

keg.data.g2$cal_abund()
keg.data.g2$cal_alphadiv()
keg.data.g2$cal_betadiv(method = methods_16s$marm$betadiv)
keg.data.g2$save_table(marmoset$microeco$data$keg$g2, sep = "\t")
keg.data.g2$save_abund(marmoset$microeco$abund$keg$g2, sep = "\t")
keg.data.g2$save_alphadiv(marmoset$microeco$alpha$keg$g2)
keg.data.g2$save_betadiv(marmoset$microeco$beta$keg$g2)
```

```{r}
keg.int$cal_tax4fun2_FRI()
```
```{r}
keg.data.int <- microtable$new(
  otu_table    = keg.int$res_tax4fun2_pathway,
  tax_table    = Tax4Fun2_KEGG$ptw_desc,
  sample_table = dataset.int$sample_table)

keg.data.int$tidy_dataset()

dataset.keg.int$sample_table <- data.frame(dataset.keg.int$sample_table, sample.table.merged[rownames(dataset.keg.int$sample_table), ])

keg.data.int$cal_abund()
keg.data.int$cal_alphadiv()
keg.data.int$cal_betadiv(method = methods_16s$marm$betadiv)
keg.data.int$save_table(marmoset$microeco$data$keg$int, sep = "\t")
keg.data.int$save_abund(marmoset$microeco$abund$keg$int, sep = "\t")
keg.data.int$save_alphadiv(marmoset$microeco$alpha$keg$int)
keg.data.int$save_betadiv(marmoset$microeco$beta$keg$int)
```
### FAPROTAX
```{r}
fpt.main        <- trans_func$new(dataset.main)
fpt.g1          <- trans_func$new(dataset.g1)
fpt.g2          <- trans_func$new(dataset.g2)
fpt.int         <- trans_func$new(dataset.int)

fpt.main$cal_spe_func(prok_database = "FAPROTAX")
fpt.g1$cal_spe_func(prok_database = "FAPROTAX")
fpt.g2$cal_spe_func(prok_database = "FAPROTAX")
fpt.int$cal_spe_func(prok_database = "FAPROTAX")

fpt.main$cal_spe_func_perc(abundance_weighted = T)
fpt.g1$cal_spe_func_perc(abundance_weighted = T)
fpt.g2$cal_spe_func_perc(abundance_weighted = T)
fpt.int$cal_spe_func_perc(abundance_weighted = T)
```
```{r}
fpt.main$trans_spe_func_perc()
fpt.g1$trans_spe_func_perc()
fpt.g2$trans_spe_func_perc()
fpt.int$trans_spe_func_perc()

fpt.main.df        <- as.data.frame(t(fpt.main$res_spe_func_perc), check.names = F)
fpt.g1.df          <- as.data.frame(t(fpt.g1$res_spe_func_perc), check.names = F)
fpt.g2.df          <- as.data.frame(t(fpt.g2$res_spe_func_perc), check.names = F)
fpt.int.df         <- as.data.frame(t(fpt.int$res_spe_func_perc), check.names = F)

fpt.main.tax        <- as.data.frame(rownames(fpt.main.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.g1.tax          <- as.data.frame(rownames(fpt.g1.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.g2.tax          <- as.data.frame(rownames(fpt.g2.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))
fpt.int.tax         <- as.data.frame(rownames(fpt.int.df), nm = c("Function")) %>% mutate(func_name = str_glue("fpt", "{row_number()}"))

fpt.main.otu        <- fpt.main.df %>% rownames_to_column("Function") %>% left_join(fpt.main.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.g1.otu          <- fpt.g1.df %>% rownames_to_column("Function") %>% left_join(fpt.g1.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.g2.otu          <- fpt.g2.df %>% rownames_to_column("Function") %>% left_join(fpt.g2.tax) %>% column_to_rownames("func_name") %>% select(-Function)
fpt.int.otu         <- fpt.int.df %>% rownames_to_column("Function") %>% left_join(fpt.int.tax) %>% column_to_rownames("func_name") %>% select(-Function)

fpt.main.tax        <- fpt.main.tax %>% column_to_rownames("func_name")
fpt.g1.tax          <- fpt.g1.tax %>% column_to_rownames("func_name")
fpt.g2.tax          <- fpt.g2.tax %>% column_to_rownames("func_name")
fpt.int.tax         <- fpt.int.tax %>% column_to_rownames("func_name")
```
```{r}
fpt.data.main <- microtable$new(otu_table    = fpt.main.otu, 
                                tax_table    = fpt.main.tax, 
                                sample_table = dataset.main$sample_table)

fpt.data.g1   <- microtable$new(otu_table    = fpt.g1.otu, 
                                tax_table    = fpt.g1.tax, 
                                sample_table = dataset.g1$sample_table)

fpt.data.g2   <- microtable$new(otu_table    = fpt.g2.otu, 
                                tax_table    = fpt.g2.tax, 
                                sample_table = dataset.g2$sample_table)

fpt.data.int  <- microtable$new(otu_table    = fpt.int.otu, 
                                tax_table    = fpt.int.tax, 
                                sample_table = dataset.int$sample_table)

fpt.data.main$tidy_dataset()
fpt.data.g1$tidy_dataset()
fpt.data.g2$tidy_dataset()
fpt.data.int$tidy_dataset()

fpt.data.main$cal_abund()
fpt.data.g1$cal_abund()
fpt.data.g2$cal_abund()
fpt.data.int$cal_abund()

fpt.data.main$cal_alphadiv()
fpt.data.g1$cal_alphadiv()
fpt.data.g2$cal_alphadiv()
fpt.data.int$cal_alphadiv()

fpt.data.main$cal_betadiv(method = methods_16s$marm$betadiv)
fpt.data.g1$cal_betadiv(method = methods_16s$marm$betadiv)
fpt.data.g2$cal_betadiv(method = methods_16s$marm$betadiv)
fpt.data.int$cal_betadiv(method = methods_16s$marm$betadiv)
```
```{r}
fpt.data.main$save_table(marmoset$microeco$data$fpt$main, sep = "\t")
fpt.data.g1$save_table(marmoset$microeco$data$fpt$g1, sep = "\t")
fpt.data.g2$save_table(marmoset$microeco$data$fpt$g2, sep = "\t")
fpt.data.int$save_table(marmoset$microeco$data$fpt$int, sep = "\t")

fpt.data.main$save_abund(marmoset$microeco$abund$fpt$main, sep = "\t")
fpt.data.g1$save_abund(marmoset$microeco$abund$fpt$g1, sep = "\t")
fpt.data.g2$save_abund(marmoset$microeco$abund$fpt$g2, sep = "\t")
fpt.data.int$save_abund(marmoset$microeco$abund$fpt$int, sep = "\t")

fpt.data.main$save_alphadiv(marmoset$microeco$alpha$fpt$main)
fpt.data.g1$save_alphadiv(marmoset$microeco$alpha$fpt$g1)
fpt.data.g2$save_alphadiv(marmoset$microeco$alpha$fpt$g2)
fpt.data.int$save_alphadiv(marmoset$microeco$alpha$fpt$int)

fpt.data.main$save_betadiv(marmoset$microeco$beta$fpt$main)
fpt.data.g1$save_betadiv(marmoset$microeco$beta$fpt$g1)
fpt.data.g2$save_betadiv(marmoset$microeco$beta$fpt$g2)
fpt.data.int$save_betadiv(marmoset$microeco$beta$fpt$int)
```
### NJC19
```{r}
njc.main        <- trans_func$new(dataset.main)
njc.g1          <- trans_func$new(dataset.g1)
njc.g2          <- trans_func$new(dataset.g2)
njc.int         <- trans_func$new(dataset.int)

njc.main$cal_spe_func(prok_database = "NJC19")
njc.g1$cal_spe_func(prok_database = "NJC19")
njc.g2$cal_spe_func(prok_database = "NJC19")
njc.int$cal_spe_func(prok_database = "NJC19")

njc.main$cal_spe_func_perc(abundance_weighted = T)
njc.g1$cal_spe_func_perc(abundance_weighted = T)
njc.g2$cal_spe_func_perc(abundance_weighted = T)
njc.int$cal_spe_func_perc(abundance_weighted = T)
```
```{r}
njc.main$trans_spe_func_perc()
njc.g1$trans_spe_func_perc()
njc.g2$trans_spe_func_perc()
njc.int$trans_spe_func_perc()

njc.main.df        <- as.data.frame(t(njc.main$res_spe_func_perc), check.names = F)
njc.g1.df          <- as.data.frame(t(njc.g1$res_spe_func_perc), check.names = F)
njc.g2.df          <- as.data.frame(t(njc.g2$res_spe_func_perc), check.names = F)
njc.int.df         <- as.data.frame(t(njc.int$res_spe_func_perc), check.names = F)

njc.main.tax        <- as.data.frame(rownames(njc.main.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.g1.tax          <- as.data.frame(rownames(njc.g1.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.g2.tax          <- as.data.frame(rownames(njc.g2.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))
njc.int.tax         <- as.data.frame(rownames(njc.int.df), nm = c("Function")) %>% mutate(func_name = str_glue("njc", "{row_number()}"))

njc.main.otu        <- njc.main.df %>% rownames_to_column("Function") %>% left_join(njc.main.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.g1.otu          <- njc.g1.df %>% rownames_to_column("Function") %>% left_join(njc.g1.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.g2.otu          <- njc.g2.df %>% rownames_to_column("Function") %>% left_join(njc.g2.tax) %>% column_to_rownames("func_name") %>% select(-Function)
njc.int.otu         <- njc.int.df %>% rownames_to_column("Function") %>% left_join(njc.int.tax) %>% column_to_rownames("func_name") %>% select(-Function)

njc.main.tax        <- njc.main.tax %>% column_to_rownames("func_name")
njc.g1.tax          <- njc.g1.tax %>% column_to_rownames("func_name")
njc.g2.tax          <- njc.g2.tax %>% column_to_rownames("func_name")
njc.int.tax         <- njc.int.tax %>% column_to_rownames("func_name")
```
```{r}
njc.data.main <- microtable$new(otu_table    = njc.main.otu, 
                                tax_table    = njc.main.tax, 
                                sample_table = dataset.main$sample_table)

njc.data.g1 <- microtable$new(otu_table    = njc.g1.otu, 
                                tax_table    = njc.g1.tax, 
                                sample_table = dataset.g1$sample_table)

njc.data.g2 <- microtable$new(otu_table    = njc.g2.otu, 
                                tax_table    = njc.g2.tax, 
                                sample_table = dataset.g2$sample_table)

njc.data.int <- microtable$new(otu_table    = njc.int.otu, 
                                tax_table    = njc.int.tax, 
                                sample_table = dataset.int$sample_table)

njc.data.main$tidy_dataset()
njc.data.g1$tidy_dataset()
njc.data.g2$tidy_dataset()
njc.data.int$tidy_dataset()

njc.data.main$cal_abund()
njc.data.g1$cal_abund()
njc.data.g2$cal_abund()
njc.data.int$cal_abund()

njc.data.main$cal_alphadiv()
njc.data.g1$cal_alphadiv()
njc.data.g2$cal_alphadiv()
njc.data.int$cal_alphadiv()

njc.data.main$cal_betadiv(method = methods_16s$marm$betadiv)
njc.data.g1$cal_betadiv(method = methods_16s$marm$betadiv)
njc.data.g2$cal_betadiv(method = methods_16s$marm$betadiv)
njc.data.int$cal_betadiv(method = methods_16s$marm$betadiv)
```
```{r}
njc.data.main$save_table(marmoset$microeco$data$njc$main, sep = "\t")
njc.data.g1$save_table(marmoset$microeco$data$njc$g1, sep = "\t")
njc.data.g2$save_table(marmoset$microeco$data$njc$g2, sep = "\t")
njc.data.int$save_table(marmoset$microeco$data$njc$int, sep = "\t")

njc.data.main$save_abund(marmoset$microeco$abund$njc$main, sep = "\t")
njc.data.g1$save_abund(marmoset$microeco$abund$njc$g1, sep = "\t")
njc.data.g2$save_abund(marmoset$microeco$abund$njc$g2, sep = "\t")
njc.data.int$save_abund(marmoset$microeco$abund$njc$int, sep = "\t")

njc.data.main$save_alphadiv(marmoset$microeco$alpha$njc$main)
njc.data.g1$save_alphadiv(marmoset$microeco$alpha$njc$g1)
njc.data.g2$save_alphadiv(marmoset$microeco$alpha$njc$g2)
njc.data.int$save_alphadiv(marmoset$microeco$alpha$njc$int)

njc.data.main$save_betadiv(marmoset$microeco$beta$njc$main)
njc.data.g1$save_betadiv(marmoset$microeco$beta$njc$g1)
njc.data.g2$save_betadiv(marmoset$microeco$beta$njc$g2)
njc.data.int$save_betadiv(marmoset$microeco$beta$njc$int)
```